# Tree Models on LSTM-Style News Features

This notebook tests whether the LSTM-style news features improve the tree models from `poc.ipynb`.

The experiment has two stages:

1. Train the three POC tree models on the full wide per-source feature set from `lstm_forecaster.ipynb`.
2. Try a reduced version of the same feature set by keeping the top news sources and grouping the rest as `Other`.

The goal is to check whether source-level news features improve next-day TA-125 direction prediction, and whether reducing sparse source columns makes the signal stronger.

Evaluation uses the statistical tests from the LSTM notebook.

In [1]:
from __future__ import annotations

import os
from typing import Any

import numpy as np
import pandas as pd
import psycopg
import requests
import yfinance as yf

import matplotlib.pyplot as plt
import seaborn as sns

from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier

from sklearn.metrics import (
    accuracy_score,
    balanced_accuracy_score,
    classification_report,
    confusion_matrix,
    roc_curve,
    auc,
)

from scipy.stats import binomtest

SEED = 42
np.random.seed(SEED)

sns.set_theme(style="whitegrid", context="notebook")
pd.set_option("display.max_columns", 80)
pd.set_option("display.width", 200)

## 1. Database Connection and Score Columns

Define the database connection, LLM score columns, and SQL helper function.

In [2]:
DB_URL = os.environ.get(
    "SENTISENSE_DATABASE_URL",
    "postgresql://sentisense:sentisense_dev@localhost:5432/sentisense",
)

SCORE_COLS = [
    "relevance_politics",
    "relevance_economy",
    "relevance_security",
    "relevance_health",
    "relevance_science",
    "relevance_technology",
    "global_sentiment",
]

def query_df(sql: str, params: tuple[Any, ...] | None = None) -> pd.DataFrame:
    """Run a SQL query and return the result as a pandas DataFrame."""
    with psycopg.connect(DB_URL) as conn, conn.cursor() as cur:
        cur.execute(sql, params or ())
        cols = [d[0] for d in cur.description]
        rows = cur.fetchall()

    return pd.DataFrame(rows, columns=cols)

print(f"DB host: {DB_URL.rsplit('@', 1)[-1]}")
print("Score columns:", SCORE_COLS)

DB host: localhost:5432/sentisense
Score columns: ['relevance_politics', 'relevance_economy', 'relevance_security', 'relevance_health', 'relevance_science', 'relevance_technology', 'global_sentiment']


## 2. Load Validated LLM-Scored Headlines

This section starts the feature pipeline from `lstm_forecaster.ipynb`.

We load validated rows from:

- `raw_headlines`
- `nlp_vectors`

Each row represents one headline with its LLM scores.

Later, we will aggregate these rows by:

- date
- news source

This is the key difference from `poc.ipynb`:
instead of using only daily averages, we keep information about which source produced which signal.

In [3]:
raw_scores = query_df("""
    SELECT rh.date::date AS date,
           rh.source,
           nv.relevance_politics,
           nv.relevance_economy,
           nv.relevance_security,
           nv.relevance_health,
           nv.relevance_science,
           nv.relevance_technology,
           nv.global_sentiment
    FROM raw_headlines rh
    JOIN nlp_vectors nv ON nv.headline_id = rh.id
    WHERE nv.validation_passed = TRUE
""")

raw_scores["date"] = pd.to_datetime(raw_scores["date"])

print(f"Loaded {len(raw_scores):,} validated rows")
print(f"Date range: {raw_scores['date'].min().date()} -> {raw_scores['date'].max().date()}")
print(f"Distinct sources: {raw_scores['source'].nunique()}")

raw_scores.head()

Loaded 1,898,499 validated rows
Date range: 2015-12-17 -> 2026-04-15
Distinct sources: 40


,date,source,relevance_politics,relevance_economy,relevance_security,relevance_health,relevance_science,relevance_technology,global_sentiment
0,2026-03-18,Ynet - חדשות,4,0,9,7,0,0,-3
1,2026-03-18,וואלה! חדשות - מבזקים,3,0,10,0,0,0,-7
2,2026-03-18,סרוגים,0,0,9,0,0,0,-5
3,2026-03-18,וואלה! חדשות - מבזקים,9,0,9,0,0,0,-3
4,2026-03-17,חדשות 0404,8,3,1,0,0,5,6


## 3. Build Wide Per-Source News Features

Aggregate headline scores by `(date, source)` and pivot them into wide daily features.

This keeps source-specific signals, unlike the simpler daily averages in `poc.ipynb`.

In [4]:
def safe_col(name: str) -> str:
    """Convert a source name into a safer column suffix."""
    return "".join(
        ch if (ch.isalnum() or ch in "_-") else "_"
        for ch in str(name)
    )

per_source_sum = (
    raw_scores
    .groupby(["date", "source"], observed=True)[SCORE_COLS]
    .sum()
    .reset_index()
)

per_source_count = (
    raw_scores
    .groupby(["date", "source"], observed=True)
    .size()
    .reset_index(name="count")
)

per_source_long = per_source_sum.merge(
    per_source_count,
    on=["date", "source"],
    how="left",
)

pivots = []

for col in [*SCORE_COLS, "count"]:
    pivot = per_source_long.pivot(
        index="date",
        columns="source",
        values=col,
    )

    pivot.columns = [
        f"{col}_{safe_col(source)}"
        for source in pivot.columns
    ]

    pivots.append(pivot)

per_source_wide = (
    pd.concat(pivots, axis=1)
    .sort_index()
    .fillna(0)
)

print("per_source_long shape:", per_source_long.shape)
print("per_source_wide shape:", per_source_wide.shape)

per_source_wide.head()

per_source_long shape: (55175, 10)
per_source_wide shape: (3771, 320)


,relevance_politics_2525_מבזקים_ותמונות,relevance_politics_IsraelDefense,relevance_politics_MivzakLive,relevance_politics_N12_-_דף_הבית,relevance_politics_N12_-_חדשות,relevance_politics_News1_-_אקטואליה,relevance_politics_News1_-_סדום_ועמורה,relevance_politics_Twoday_-_חדשות,relevance_politics_Ynet_-_חדשות,relevance_politics_Ynet_-_מבזקי_חדשות,relevance_politics_nrg_-_חדשות,relevance_politics_nrg_-_מבזקים,relevance_politics_גלי_צה_ל,relevance_politics_גלי_צה_ל_-_מבזקים,relevance_politics_דבר_-_חדשות,relevance_politics_דבר_-_עבודה,relevance_politics_הארץ_-_חדשות,relevance_politics_הארץ_-_מבזקים,relevance_politics_וואלה__חדשות,relevance_politics_וואלה__חדשות_-_מבזקים,relevance_politics_זמן_ישראל,relevance_politics_חדשות_0404,relevance_politics_חדשות_13,relevance_politics_חדשות_נת_ע,relevance_politics_ישראל_היום_-_העולם,relevance_politics_ישראל_היום_-_חדשות,relevance_politics_כיפה_-_חדשות,relevance_politics_מעריב,relevance_politics_מעריב_-_מבזקים,relevance_politics_מקור_ראשון_-_בעולם,relevance_politics_מקור_ראשון_-_חדשות,relevance_politics_סרוגים,relevance_politics_עכשיו_14_-_חדשות_ואקטואליה,relevance_politics_עכשיו_14_-_מבזקים,relevance_politics_עכשיו_14_-_פוליטי_מדיני,relevance_politics_ערוץ_20_-_חדשות,relevance_politics_ערוץ_7_-_מבזקים,relevance_politics_רשת_-_חדשות,relevance_politics_רשת_ב_,relevance_politics_תיקדבקה,...,count_2525_מבזקים_ותמונות,count_IsraelDefense,count_MivzakLive,count_N12_-_דף_הבית,count_N12_-_חדשות,count_News1_-_אקטואליה,count_News1_-_סדום_ועמורה,count_Twoday_-_חדשות,count_Ynet_-_חדשות,count_Ynet_-_מבזקי_חדשות,count_nrg_-_חדשות,count_nrg_-_מבזקים,count_גלי_צה_ל,count_גלי_צה_ל_-_מבזקים,count_דבר_-_חדשות,count_דבר_-_עבודה,count_הארץ_-_חדשות,count_הארץ_-_מבזקים,count_וואלה__חדשות,count_וואלה__חדשות_-_מבזקים,count_זמן_ישראל,count_חדשות_0404,count_חדשות_13,count_חדשות_נת_ע,count_ישראל_היום_-_העולם,count_ישראל_היום_-_חדשות,count_כיפה_-_חדשות,count_מעריב,count_מעריב_-_מבזקים,count_מקור_ראשון_-_בעולם,count_מקור_ראשון_-_חדשות,count_סרוגים,count_עכשיו_14_-_חדשות_ואקטואליה,count_עכשיו_14_-_מבזקים,count_עכשיו_14_-_פוליטי_מדיני,count_ערוץ_20_-_חדשות,count_ערוץ_7_-_מבזקים,count_רשת_-_חדשות,count_רשת_ב_,count_תיקדבקה
date,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
2015-12-17,13.0,0.0,0.0,119.0,264.0,0.0,0.0,0.0,0.0,0.0,258.0,254.0,92.0,188.0,0.0,0.0,143.0,202.0,0.0,0.0,0.0,0.0,161.0,0.0,27.0,0.0,58.0,189.0,0.0,0.0,0.0,196.0,0.0,0.0,0.0,0.0,0.0,0.0,178.0,81.0,...,3.0,0.0,0.0,32.0,71.0,1.0,0.0,0.0,0.0,0.0,52.0,50.0,15.0,39.0,0.0,0.0,41.0,36.0,0.0,0.0,0.0,0.0,32.0,0.0,13.0,0.0,10.0,56.0,0.0,0.0,0.0,32.0,0.0,0.0,0.0,0.0,0.0,0.0,46.0,14.0
2015-12-18,9.0,0.0,0.0,57.0,164.0,0.0,0.0,0.0,0.0,0.0,154.0,72.0,43.0,122.0,0.0,0.0,111.0,251.0,0.0,0.0,0.0,0.0,137.0,0.0,9.0,0.0,25.0,175.0,0.0,0.0,0.0,52.0,0.0,0.0,0.0,0.0,0.0,0.0,74.0,25.0,...,1.0,0.0,0.0,20.0,38.0,0.0,0.0,0.0,0.0,0.0,27.0,18.0,6.0,25.0,0.0,0.0,23.0,41.0,0.0,0.0,0.0,0.0,28.0,0.0,4.0,0.0,3.0,44.0,0.0,0.0,0.0,8.0,0.0,0.0,0.0,0.0,0.0,0.0,18.0,5.0
2015-12-19,10.0,0.0,0.0,54.0,101.0,0.0,0.0,0.0,0.0,0.0,81.0,36.0,29.0,82.0,0.0,0.0,64.0,161.0,0.0,0.0,0.0,0.0,93.0,0.0,4.0,0.0,2.0,193.0,0.0,0.0,0.0,51.0,0.0,0.0,0.0,0.0,0.0,0.0,124.0,30.0,...,1.0,0.0,0.0,18.0,28.0,0.0,0.0,0.0,0.0,0.0,13.0,15.0,4.0,16.0,0.0,0.0,20.0,29.0,0.0,0.0,0.0,0.0,16.0,0.0,2.0,0.0,2.0,49.0,0.0,0.0,0.0,10.0,0.0,0.0,0.0,0.0,0.0,0.0,35.0,8.0
2015-12-20,81.0,0.0,0.0,150.0,311.0,0.0,0.0,0.0,0.0,0.0,324.0,216.0,81.0,144.0,0.0,0.0,175.0,365.0,0.0,0.0,0.0,0.0,136.0,0.0,57.0,0.0,47.0,281.0,0.0,0.0,0.0,194.0,0.0,0.0,0.0,0.0,0.0,0.0,74.0,53.0,...,15.0,0.0,0.0,30.0,71.0,0.0,0.0,0.0,0.0,0.0,52.0,58.0,14.0,42.0,0.0,0.0,36.0,68.0,0.0,0.0,0.0,0.0,28.0,0.0,12.0,0.0,7.0,62.0,0.0,0.0,0.0,35.0,0.0,0.0,0.0,0.0,0.0,0.0,30.0,14.0
2015-12-21,82.0,0.0,0.0,140.0,238.0,0.0,0.0,0.0,0.0,0.0,351.0,336.0,58.0,202.0,0.0,0.0,125.0,292.0,0.0,0.0,0.0,0.0,203.0,0.0,12.0,0.0,90.0,294.0,0.0,0.0,0.0,216.0,0.0,0.0,0.0,0.0,0.0,0.0,90.0,38.0,...,11.0,0.0,0.0,28.0,59.0,0.0,0.0,0.0,0.0,0.0,62.0,78.0,9.0,43.0,0.0,0

## 4. Load Market and Finance Data

The target is based on the TA-125 index, so TA-125 will define the trading-day calendar.

Load the financial data used alongside the news features:

- TA-125
- VTA-35
- S&P 500
- VIX
- Brent Oil
- USD/ILS

In [5]:
def convert_volume(val: Any) -> float:
    """Convert values like '1.2M', '850K', or '1,234' into numeric volume."""
    if pd.isna(val):
        return np.nan

    s = str(val).upper().replace(",", "").strip()

    if s.endswith("M"):
        return float(s[:-1]) * 1_000_000
    if s.endswith("B"):
        return float(s[:-1]) * 1_000_000_000
    if s.endswith("K"):
        return float(s[:-1]) * 1_000

    return float(s)


def to_float_series(s: pd.Series) -> pd.Series:
    """Convert comma-formatted numeric strings into floats."""
    if pd.api.types.is_numeric_dtype(s):
        return s.astype(float)

    return s.astype(str).str.replace(",", "", regex=False).astype(float)


def gather_market_data(start="2015-12-17", end=None) -> pd.DataFrame:
    """Download S&P 500, VIX, and Brent Oil close prices from Yahoo Finance."""
    end = end or pd.Timestamp.today().strftime("%Y-%m-%d")

    tickers = {
        "SP500": "^GSPC",
        "VIX": "^VIX",
        "Brent_Oil": "BZ=F",
    }

    data = yf.download(
        list(tickers.values()),
        start=start,
        end=end,
        progress=False,
    )

    market = data["Close"].rename(columns={v: k for k, v in tickers.items()})
    market.index = pd.to_datetime(market.index)

    return market.add_prefix("Market_")


def get_usd_ils_history(start="2015-12-17") -> pd.DataFrame:
    """Download historical USD/ILS exchange rates from Frankfurter."""
    url = f"https://api.frankfurter.app/{start}..?from=USD&to=ILS"

    response = requests.get(url, timeout=30)
    response.raise_for_status()

    rates = response.json()["rates"]

    fx = pd.DataFrame.from_dict(rates, orient="index")
    fx.index = pd.to_datetime(fx.index)
    fx.columns = ["FX_USD_ILS"]

    return fx.sort_index()


TA125_PATH = "TA 125 Historical Data.csv"
VTA35_PATH = "Tel Aviv Volatility Index VTA35 Historical Data.csv"

ta125 = pd.read_csv(TA125_PATH)
vta35 = pd.read_csv(VTA35_PATH)

market = gather_market_data()
fx = get_usd_ils_history()

print("TA125:", ta125.shape)
print("VTA35:", vta35.shape)
print("Market:", market.shape)
print("FX:", fx.shape)

TA125: (2543, 7)
VTA35: (1670, 7)
Market: (2635, 3)
FX: (2678, 1)


## 5. Clean TA-125 and VTA-35 Data

The local CSV files contain several columns, but for this experiment we keep only the core fields:

- `TA125_Price`: used to create the prediction target
- `TA125_Volume`: market-volume feature
- `VTA35_Price`: Israeli volatility feature

We also convert comma-formatted strings and volume suffixes (`K`, `M`, `B`) into numeric values.

In [6]:
ta125_clean = ta125.copy()
ta125_clean["Date"] = pd.to_datetime(ta125_clean["Date"])
ta125_clean = ta125_clean.set_index("Date").sort_index()

ta125_clean = ta125_clean.rename(columns={
    "Price": "TA125_Price",
    "Vol.": "TA125_Volume",
})

ta125_clean["TA125_Price"] = to_float_series(ta125_clean["TA125_Price"])
ta125_clean["TA125_Volume"] = ta125_clean["TA125_Volume"].apply(convert_volume)

ta125_clean = ta125_clean[["TA125_Price", "TA125_Volume"]]


vta35_clean = vta35.copy()
vta35_clean["Date"] = pd.to_datetime(vta35_clean["Date"])
vta35_clean = vta35_clean.set_index("Date").sort_index()

vta35_clean = vta35_clean.rename(columns={
    "Price": "VTA35_Price",
})

vta35_clean["VTA35_Price"] = to_float_series(vta35_clean["VTA35_Price"])
vta35_clean = vta35_clean[["VTA35_Price"]]


print("TA125 clean:")
print(ta125_clean.head())
print()
print("VTA35 clean:")
print(vta35_clean.head())

print()
print("TA125 clean shape:", ta125_clean.shape)
print("VTA35 clean shape:", vta35_clean.shape)

TA125 clean:
            TA125_Price  TA125_Volume
Date                                 
2015-12-17      1311.13   128130000.0
2015-12-20      1293.30    79590000.0
2015-12-21      1296.50    86200000.0
2015-12-22      1287.48   101590000.0
2015-12-23      1294.11   126130000.0

VTA35 clean:
            VTA35_Price
Date                   
2019-07-17        13.54
2019-07-18        13.90
2019-07-21        14.16
2019-07-22        12.96
2019-07-23        11.72

TA125 clean shape: (2543, 2)
VTA35 clean shape: (1670, 1)


In [7]:
def add_vta35_availability(df: pd.DataFrame) -> pd.DataFrame:
    out = df.copy()

    vta35_cols = [col for col in out.columns if col.startswith("VTA35_")]

    if vta35_cols:
        out["VTA35_available"] = out[vta35_cols].notna().any(axis=1).astype(int)
        out[vta35_cols] = out[vta35_cols].fillna(0)

    return out

## 6. Align News Features to Trading Days

Attach each news date to the next available TA-125 trading day.

This keeps weekend news and matches the calendar logic from `lstm_forecaster.ipynb`.

In [8]:
trading_days = pd.DatetimeIndex(ta125_clean.index).sort_values()

trading_arr = np.asarray(trading_days)
news_dates = per_source_wide.index.values

positions = np.searchsorted(trading_arr, news_dates, side="left")
mask = positions < len(trading_arr)

news_attached = per_source_wide.iloc[mask].copy()
news_attached["trading_day"] = trading_days[positions[mask]]

news_per_trading_day = (
    news_attached
    .groupby("trading_day")
    .sum()
    .sort_index()
)

print("Original news dates:", per_source_wide.shape)
print("News attached to trading days:", news_per_trading_day.shape)

news_per_trading_day.head()

Original news dates: (3771, 320)
News attached to trading days: (2534, 320)


,relevance_politics_2525_מבזקים_ותמונות,relevance_politics_IsraelDefense,relevance_politics_MivzakLive,relevance_politics_N12_-_דף_הבית,relevance_politics_N12_-_חדשות,relevance_politics_News1_-_אקטואליה,relevance_politics_News1_-_סדום_ועמורה,relevance_politics_Twoday_-_חדשות,relevance_politics_Ynet_-_חדשות,relevance_politics_Ynet_-_מבזקי_חדשות,relevance_politics_nrg_-_חדשות,relevance_politics_nrg_-_מבזקים,relevance_politics_גלי_צה_ל,relevance_politics_גלי_צה_ל_-_מבזקים,relevance_politics_דבר_-_חדשות,relevance_politics_דבר_-_עבודה,relevance_politics_הארץ_-_חדשות,relevance_politics_הארץ_-_מבזקים,relevance_politics_וואלה__חדשות,relevance_politics_וואלה__חדשות_-_מבזקים,relevance_politics_זמן_ישראל,relevance_politics_חדשות_0404,relevance_politics_חדשות_13,relevance_politics_חדשות_נת_ע,relevance_politics_ישראל_היום_-_העולם,relevance_politics_ישראל_היום_-_חדשות,relevance_politics_כיפה_-_חדשות,relevance_politics_מעריב,relevance_politics_מעריב_-_מבזקים,relevance_politics_מקור_ראשון_-_בעולם,relevance_politics_מקור_ראשון_-_חדשות,relevance_politics_סרוגים,relevance_politics_עכשיו_14_-_חדשות_ואקטואליה,relevance_politics_עכשיו_14_-_מבזקים,relevance_politics_עכשיו_14_-_פוליטי_מדיני,relevance_politics_ערוץ_20_-_חדשות,relevance_politics_ערוץ_7_-_מבזקים,relevance_politics_רשת_-_חדשות,relevance_politics_רשת_ב_,relevance_politics_תיקדבקה,...,count_2525_מבזקים_ותמונות,count_IsraelDefense,count_MivzakLive,count_N12_-_דף_הבית,count_N12_-_חדשות,count_News1_-_אקטואליה,count_News1_-_סדום_ועמורה,count_Twoday_-_חדשות,count_Ynet_-_חדשות,count_Ynet_-_מבזקי_חדשות,count_nrg_-_חדשות,count_nrg_-_מבזקים,count_גלי_צה_ל,count_גלי_צה_ל_-_מבזקים,count_דבר_-_חדשות,count_דבר_-_עבודה,count_הארץ_-_חדשות,count_הארץ_-_מבזקים,count_וואלה__חדשות,count_וואלה__חדשות_-_מבזקים,count_זמן_ישראל,count_חדשות_0404,count_חדשות_13,count_חדשות_נת_ע,count_ישראל_היום_-_העולם,count_ישראל_היום_-_חדשות,count_כיפה_-_חדשות,count_מעריב,count_מעריב_-_מבזקים,count_מקור_ראשון_-_בעולם,count_מקור_ראשון_-_חדשות,count_סרוגים,count_עכשיו_14_-_חדשות_ואקטואליה,count_עכשיו_14_-_מבזקים,count_עכשיו_14_-_פוליטי_מדיני,count_ערוץ_20_-_חדשות,count_ערוץ_7_-_מבזקים,count_רשת_-_חדשות,count_רשת_ב_,count_תיקדבקה
trading_day,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
2015-12-17,13.0,0.0,0.0,119.0,264.0,0.0,0.0,0.0,0.0,0.0,258.0,254.0,92.0,188.0,0.0,0.0,143.0,202.0,0.0,0.0,0.0,0.0,161.0,0.0,27.0,0.0,58.0,189.0,0.0,0.0,0.0,196.0,0.0,0.0,0.0,0.0,0.0,0.0,178.0,81.0,...,3.0,0.0,0.0,32.0,71.0,1.0,0.0,0.0,0.0,0.0,52.0,50.0,15.0,39.0,0.0,0.0,41.0,36.0,0.0,0.0,0.0,0.0,32.0,0.0,13.0,0.0,10.0,56.0,0.0,0.0,0.0,32.0,0.0,0.0,0.0,0.0,0.0,0.0,46.0,14.0
2015-12-20,100.0,0.0,0.0,261.0,576.0,0.0,0.0,0.0,0.0,0.0,559.0,324.0,153.0,348.0,0.0,0.0,350.0,777.0,0.0,0.0,0.0,0.0,366.0,0.0,70.0,0.0,74.0,649.0,0.0,0.0,0.0,297.0,0.0,0.0,0.0,0.0,0.0,0.0,272.0,108.0,...,17.0,0.0,0.0,68.0,137.0,0.0,0.0,0.0,0.0,0.0,92.0,91.0,24.0,83.0,0.0,0.0,79.0,138.0,0.0,0.0,0.0,0.0,72.0,0.0,18.0,0.0,12.0,155.0,0.0,0.0,0.0,53.0,0.0,0.0,0.0,0.0,0.0,0.0,83.0,27.0
2015-12-21,82.0,0.0,0.0,140.0,238.0,0.0,0.0,0.0,0.0,0.0,351.0,336.0,58.0,202.0,0.0,0.0,125.0,292.0,0.0,0.0,0.0,0.0,203.0,0.0,12.0,0.0,90.0,294.0,0.0,0.0,0.0,216.0,0.0,0.0,0.0,0.0,0.0,0.0,90.0,38.0,...,11.0,0.0,0.0,28.0,59.0,0.0,0.0,0.0,0.0,0.0,62.0,78.0,9.0,43.0,0.0,0.0,34.0,44.0,0.0,0.0,0.0,0.0,36.0,0.0,7.0,0.0,12.0,66.0,0.0,0.0,0.0,38.0,0.0,0.0,0.0,0.0,0.0,0.0,20.0,9.0
2015-12-22,65.0,0.0,0.0,138.0,274.0,0.0,3.0,0.0,0.0,0.0,339.0,275.0,67.0,189.0,0.0,0.0,175.0,317.0,0.0,0.0,0.0,0.0,200.0,0.0,33.0,0.0,69.0,309.0,0.0,0.0,0.0,191.0,0.0,0.0,0.0,0.0,0.0,0.0,200.0,41.0,...,9.0,0.0,0.0,30.0,68.0,0.0,1.0,0.0,0.0,0.0,53.0,49.0,13.0,36.0,0.0,0.0,35.0,57.0,0.0,0.0,0.0,0.0,38.0,0.0,10.0,0.0,11.0,63.0,0.0,0.0,0.0,31.0,0.0,0.0,0.0,0.0,0.0,0.0,50.0,5.0
2015-12-23,43.0,0.0,0.0,108.0,215.0,0.0,9.0,0.0,0.0,0.0,297.0,189.0,54.0,170.0,0.0,0.0,129.0,239.0,0.0,0.0,0.0,0.0,154.0,0.0,35.0,0.0,83.0,171.0,0.0,0.0,0.0,245.0,0.0,0.0,0.0,0.0,0.0,0.0,209.0,69.0,...,7.0,0.0,0.0,25.0,63.0,0.0,1.0,0

## 7. Merge Feature Blocks

In [9]:
merged = pd.DataFrame(index=trading_days)
merged.index.name = "date"

merged = (
    merged
    .join(ta125_clean, how="left")
    .join(vta35_clean, how="left")
    .join(market, how="left")
    .join(fx, how="left")
    .join(news_per_trading_day, how="left")
)

market_like_cols = [
    col for col in merged.columns
    if col.startswith(("Market_", "FX_", "VTA35_"))
]

news_cols = news_per_trading_day.columns.tolist()

merged[market_like_cols] = merged[market_like_cols].ffill()
merged[news_cols] = merged[news_cols].fillna(0)

print("Merged shape:", merged.shape)
print("Date range:", merged.index.min().date(), "->", merged.index.max().date())
print("Missing values by block:")
print("TA125 missing:", merged[["TA125_Price", "TA125_Volume"]].isna().sum().sum())
print("Market/FX/VTA35 missing:", merged[market_like_cols].isna().sum().sum())
print("News missing:", merged[news_cols].isna().sum().sum())

merged.head()

Merged shape: (2543, 327)
Date range: 2015-12-17 -> 2026-04-29
Missing values by block:
TA125 missing: 26
Market/FX/VTA35 missing: 881
News missing: 0


,TA125_Price,TA125_Volume,VTA35_Price,Market_Brent_Oil,Market_SP500,Market_VIX,FX_USD_ILS,relevance_politics_2525_מבזקים_ותמונות,relevance_politics_IsraelDefense,relevance_politics_MivzakLive,relevance_politics_N12_-_דף_הבית,relevance_politics_N12_-_חדשות,relevance_politics_News1_-_אקטואליה,relevance_politics_News1_-_סדום_ועמורה,relevance_politics_Twoday_-_חדשות,relevance_politics_Ynet_-_חדשות,relevance_politics_Ynet_-_מבזקי_חדשות,relevance_politics_nrg_-_חדשות,relevance_politics_nrg_-_מבזקים,relevance_politics_גלי_צה_ל,relevance_politics_גלי_צה_ל_-_מבזקים,relevance_politics_דבר_-_חדשות,relevance_politics_דבר_-_עבודה,relevance_politics_הארץ_-_חדשות,relevance_politics_הארץ_-_מבזקים,relevance_politics_וואלה__חדשות,relevance_politics_וואלה__חדשות_-_מבזקים,relevance_politics_זמן_ישראל,relevance_politics_חדשות_0404,relevance_politics_חדשות_13,relevance_politics_חדשות_נת_ע,relevance_politics_ישראל_היום_-_העולם,relevance_politics_ישראל_היום_-_חדשות,relevance_politics_כיפה_-_חדשות,relevance_politics_מעריב,relevance_politics_מעריב_-_מבזקים,relevance_politics_מקור_ראשון_-_בעולם,relevance_politics_מקור_ראשון_-_חדשות,relevance_politics_סרוגים,relevance_politics_עכשיו_14_-_חדשות_ואקטואליה,...,count_2525_מבזקים_ותמונות,count_IsraelDefense,count_MivzakLive,count_N12_-_דף_הבית,count_N12_-_חדשות,count_News1_-_אקטואליה,count_News1_-_סדום_ועמורה,count_Twoday_-_חדשות,count_Ynet_-_חדשות,count_Ynet_-_מבזקי_חדשות,count_nrg_-_חדשות,count_nrg_-_מבזקים,count_גלי_צה_ל,count_גלי_צה_ל_-_מבזקים,count_דבר_-_חדשות,count_דבר_-_עבודה,count_הארץ_-_חדשות,count_הארץ_-_מבזקים,count_וואלה__חדשות,count_וואלה__חדשות_-_מבזקים,count_זמן_ישראל,count_חדשות_0404,count_חדשות_13,count_חדשות_נת_ע,count_ישראל_היום_-_העולם,count_ישראל_היום_-_חדשות,count_כיפה_-_חדשות,count_מעריב,count_מעריב_-_מבזקים,count_מקור_ראשון_-_בעולם,count_מקור_ראשון_-_חדשות,count_סרוגים,count_עכשיו_14_-_חדשות_ואקטואליה,count_עכשיו_14_-_מבזקים,count_עכשיו_14_-_פוליטי_מדיני,count_ערוץ_20_-_חדשות,count_ערוץ_7_-_מבזקים,count_רשת_-_חדשות,count_רשת_ב_,count_תיקדבקה
date,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
2015-12-17,1311.13,128130000.0,NaN,37.180000,2041.890015,18.940001,3.8926,13.0,0.0,0.0,119.0,264.0,0.0,0.0,0.0,0.0,0.0,258.0,254.0,92.0,188.0,0.0,0.0,143.0,202.0,0.0,0.0,0.0,0.0,161.0,0.0,27.0,0.0,58.0,189.0,0.0,0.0,0.0,196.0,0.0,...,3.0,0.0,0.0,32.0,71.0,1.0,0.0,0.0,0.0,0.0,52.0,50.0,15.0,39.0,0.0,0.0,41.0,36.0,0.0,0.0,0.0,0.0,32.0,0.0,13.0,0.0,10.0,56.0,0.0,0.0,0.0,32.0,0.0,0.0,0.0,0.0,0.0,0.0,46.0,14.0
2015-12-20,1293.30,79590000.0,NaN,37.180000,2041.890015,18.940001,3.8926,100.0,0.0,0.0,261.0,576.0,0.0,0.0,0.0,0.0,0.0,559.0,324.0,153.0,348.0,0.0,0.0,350.0,777.0,0.0,0.0,0.0,0.0,366.0,0.0,70.0,0.0,74.0,649.0,0.0,0.0,0.0,297.0,0.0,...,17.0,0.0,0.0,68.0,137.0,0.0,0.0,0.0,0.0,0.0,92.0,91.0,24.0,83.0,0.0,0.0,79.0,138.0,0.0,0.0,0.0,0.0,72.0,0.0,18.0,0.0,12.0,155.0,0.0,0.0,0.0,53.0,0.0,0.0,0.0,0.0,0.0,0.0,83.0,27.0
2015-12-21,1296.50,86200000.0,NaN,36.349998,2021.150024,18.700001,3.9021,82.0,0.0,0.0,140.0,238.0,0.0,0.0,0.0,0.0,0.0,351.0,336.0,58.0,202.0,0.0,0.0,125.0,292.0,0.0,0.0,0.0,0.0,203.0,0.0,12.0,0.0,90.0,294.0,0.0,0.0,0.0,216.0,0.0,...,11.0,0.0,0.0,28.0,59.0,0.0,0.0,0.0,0.0,0.0,62.0,78.0,9.0,43.0,0.0,0.0,34.0,44.0,0.0,0.0,0.0,0.0,36.0,0.0,7.0,0.0,12.0,66.0,0.0,0.0,0.0,38.0,0.0,0.0,0.0,0.0,0.0,0.0,20.0,9.0
2015-12-22,1287.48,101590000.0,NaN,36.110001,2038.969971,16.600000,3.8988,65.0,0.0,0.0,138.0,274.0,0.0,3.0,0.0,0.0,0.0,339.0,275.0,67.0,189.0,0.0,0.0,175.0,317.0,0.0,0.0,0.0,0.0,200.0,0.0,33.0,0.0,69.0,309.0,0.0,0.0,0.0,191.0,0.0,...,9.0,0.0,0.0,30.0,68.0,0.0,1.0,0.0,0.0,0.0,53.0,49.0,13.0,36.0,0.0,0.0,35.0,57.0,0.0,0.0,0.0,0.0,38.0,0.0,10.0,0.0,11.0,63.0,0.0,0.0,0.0,31.0,0.0,0.0,0.0,0.0,0.0,0.0,50.0,5.0
2015-12-23,1294.11,126130000.0,NaN,37.360001,2064.290039,15.570000,3.8920,43.0,0.0,0.0,108.0,215.0,0.0,9.0,0.0,0.0,0.0,297.0,189.0,54.0,170.0,0.0,0.0,129.0,239.0,0.0,0.0,0.0,0.0,154.0,0.0,35.0,0.0,83.0,171.0,0.0,0.0,0.0,245.0,0.0,...,7.0,0.0,0.0,25.0,63.0,0.0

## 8. Create the Prediction Target

The prediction target is:

> `1` if the TA-125 index rises on the next trading day, otherwise `0`.

The dataset is sorted from oldest to newest, so the next trading day is represented by `shift(-1)`.

After creating the target, we drop `TA125_Price`.

This prevents the model from using the raw current index price as a shortcut feature.

In [10]:
merged = merged.sort_index().copy()

next_ta125_price = merged["TA125_Price"].shift(-1)

merged["Target"] = np.where(
    next_ta125_price.isna(),
    np.nan,
    (next_ta125_price > merged["TA125_Price"]).astype(int),
)

merged = merged.drop(columns=["TA125_Price"])

merged = add_vta35_availability(merged)

df_ml = (
    merged
    .replace([np.inf, -np.inf], np.nan)
    .dropna()
    .copy()
)

X = df_ml.drop(columns=["Target"])
y = df_ml["Target"].astype(int)

print("Final dataset shape:", df_ml.shape)
print("Feature count:", X.shape[1])
print("Date range:", df_ml.index.min().date(), "->", df_ml.index.max().date())
print()
print("Target balance:")
print(y.value_counts(normalize=True).rename({0: "Fall", 1: "Rise"}))

print()
print("VTA35 availability:")
print(df_ml["VTA35_available"].value_counts())

Final dataset shape: (2516, 328)
Feature count: 327
Date range: 2015-12-17 -> 2026-04-28

Target balance:
Target
Rise    0.538553
Fall    0.461447
Name: proportion, dtype: float64

VTA35 availability:
VTA35_available
1    1655
0     861
Name: count, dtype: int64


## 9. Chronological Train/Test Split

In [11]:
split_idx = int(len(df_ml) * 0.8)

X_train = X.iloc[:split_idx]
X_test = X.iloc[split_idx:]

y_train = y.iloc[:split_idx]
y_test = y.iloc[split_idx:]

print(f"Train rows: {len(X_train):,}")
print(f"Test rows:  {len(X_test):,}")
print()
print(f"Train date range: {X_train.index.min().date()} -> {X_train.index.max().date()}")
print(f"Test date range:  {X_test.index.min().date()} -> {X_test.index.max().date()}")
print()
print(f"Train rise rate: {y_train.mean():.2%}")
print(f"Test rise rate:  {y_test.mean():.2%}")

Train rows: 2,012
Test rows:  504

Train date range: 2015-12-17 -> 2024-03-25
Test date range:  2024-03-26 -> 2026-04-28

Train rise rate: 53.13%
Test rise rate:  56.75%


## 10. Train POC Tree Models on LSTM Features

Now we train the three classical models from `poc.ipynb`:

- XGBoost
- LightGBM
- CatBoost

The important difference is the input data:

In `poc.ipynb`, these models used simple daily-mean news features.

Here, they use the wide per-source feature set from `lstm_forecaster.ipynb`.

This lets us test whether the LSTM feature representation is useful even without an LSTM model.

In [12]:
models = {
    "XGBoost": XGBClassifier(
        n_estimators=200,
        max_depth=4,
        learning_rate=0.05,
        subsample=0.8,
        colsample_bytree=0.8,
        eval_metric="logloss",
        random_state=SEED,
    ),
    "LGBM": LGBMClassifier(
        n_estimators=200,
        learning_rate=0.03,
        num_leaves=31,
        subsample=0.8,
        colsample_bytree=0.8,
        random_state=SEED,
        verbose=-1,
    ),
    "CatBoost": CatBoostClassifier(
        iterations=200,
        learning_rate=0.03,
        depth=6,
        l2_leaf_reg=3,
        verbose=0,
        random_seed=SEED,
    ),
}

predictions = {}

for name, model in models.items():
    print(f"Training {name}...")

    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)
    y_proba = model.predict_proba(X_test)[:, 1]

    predictions[name] = {
        "model": model,
        "y_pred": y_pred,
        "y_proba": y_proba,
    }

    print(f"{name} accuracy: {accuracy_score(y_test, y_pred):.2%}")
    print(classification_report(y_test, y_pred, target_names=["Fall", "Rise"]))
    print("-" * 80)

Training XGBoost...
XGBoost accuracy: 56.94%
              precision    recall  f1-score   support

        Fall       0.53      0.05      0.08       218
        Rise       0.57      0.97      0.72       286

    accuracy                           0.57       504
   macro avg       0.55      0.51      0.40       504
weighted avg       0.55      0.57      0.44       504

--------------------------------------------------------------------------------
Training LGBM...
LGBM accuracy: 56.94%
              precision    recall  f1-score   support

        Fall       0.53      0.04      0.08       218
        Rise       0.57      0.97      0.72       286

    accuracy                           0.57       504
   macro avg       0.55      0.51      0.40       504
weighted avg       0.55      0.57      0.44       504

--------------------------------------------------------------------------------
Training CatBoost...
CatBoost accuracy: 57.14%
              precision    recall  f1-score   support

## 11. Statistical Evaluation

Compare each model against the majority-class baseline using accuracy,
balanced accuracy, binomial test, permutation test, bootstrap CI, and ROC/AUC.

In [13]:
baseline_class = int(y_train.mean() > 0.5)
baseline_acc = float((y_test == baseline_class).mean())

print(f"Majority baseline class: {baseline_class}")
print(f"Majority baseline accuracy: {baseline_acc:.2%}")

Majority baseline class: 1
Majority baseline accuracy: 56.75%


In [14]:
rng = np.random.default_rng(SEED)

rows = []

y_test_arr = y_test.values
n_test = len(y_test_arr)

for name, pred_data in predictions.items():
    y_pred = pred_data["y_pred"]
    y_proba = pred_data["y_proba"]

    acc = float(accuracy_score(y_test_arr, y_pred))
    bal_acc = float(balanced_accuracy_score(y_test_arr, y_pred))
    n_correct = int((y_pred == y_test_arr).sum())

    # Binomial test: is the model significantly better than the majority baseline?
    p_binom = binomtest(
        n_correct,
        n_test,
        p=baseline_acc,
        alternative="greater",
    ).pvalue

    # Bootstrap confidence interval for accuracy
    correct_mask = y_pred == y_test_arr

    boot_accs = np.empty(1000)
    for k in range(1000):
        idx = rng.integers(0, n_test, n_test)
        boot_accs[k] = correct_mask[idx].mean()

    ci_low, ci_high = np.percentile(boot_accs, [2.5, 97.5])

    # Permutation test: compare model accuracy against shuffled labels
    perm_accs = np.empty(1000)
    for k in range(1000):
        shuffled = rng.permutation(y_test_arr)
        perm_accs[k] = (y_pred == shuffled).mean()

    p_perm = float((perm_accs >= acc).mean())

    fpr, tpr, _ = roc_curve(y_test_arr, y_proba)
    roc_auc = float(auc(fpr, tpr))

    rows.append({
        "Model": name,
        "Accuracy": acc,
        "Balanced Accuracy": bal_acc,
        "Baseline": baseline_acc,
        "Gap vs Baseline": acc - baseline_acc,
        "p_binom": p_binom,
        "p_perm": p_perm,
        "CI_low": ci_low,
        "CI_high": ci_high,
        "ROC_AUC": roc_auc,
    })

results_df = (
    pd.DataFrame(rows)
    .sort_values("Accuracy", ascending=False)
    .reset_index(drop=True)
)

results_df

,Model,Accuracy,Balanced Accuracy,Baseline,Gap vs Baseline,p_binom,p_perm,CI_low,CI_high,ROC_AUC
0,CatBoost,0.571429,0.506768,0.56746,0.003968,0.447154,0.234,0.527778,0.613095,0.475974
1,XGBoost,0.569444,0.507202,0.56746,0.001984,0.482880,0.255,0.525794,0.611161,0.519592
2,LGBM,0.569444,0.506656,0.56746,0.001984,0.482880,0.278,0.525794,0.615079,0.545294


## 12. Compare Against Existing Notebooks

Compare the new results against:

- `poc.ipynb`: tree models with daily-mean features
- `lstm_forecaster.ipynb`: LSTM with wide per-source features

This shows whether the improvement comes from the features, the model type, or neither.

## 13. Experiment Plan

Test one change at a time:
aggregation method, feature set, model family, and tuning strategy.
Every run is evaluated with the same statistical tests.

In [15]:
experiment_results = results_df.copy()

experiment_results.insert(0, "Experiment", "Baseline wide features")
experiment_results.insert(1, "Aggregation", "LSTM per-source wide sum")
experiment_results.insert(2, "Feature Set", "News per source + market basics")
experiment_results.insert(3, "Tuning", "POC default parameters")

experiment_results

,Experiment,Aggregation,Feature Set,Tuning,Model,Accuracy,Balanced Accuracy,Baseline,Gap vs Baseline,p_binom,p_perm,CI_low,CI_high,ROC_AUC
0,Baseline wide features,LSTM per-source wide sum,News per source + market basics,POC default parameters,CatBoost,0.571429,0.506768,0.56746,0.003968,0.447154,0.234,0.527778,0.613095,0.475974
1,Baseline wide features,LSTM per-source wide sum,News per source + market basics,POC default parameters,XGBoost,0.569444,0.507202,0.56746,0.001984,0.482880,0.255,0.525794,0.611161,0.519592
2,Baseline wide features,LSTM per-source wide sum,News per source + market basics,POC default parameters,LGBM,0.569444,0.506656,0.56746,0.001984,0.482880,0.278,0.525794,0.615079,0.545294


## 14. Improvement Attempt 1: Top Sources + Other

The full wide feature set has 320 news columns from 40 sources, which may be too sparse for the available training data.

To reduce noise, we keep only the top `N` sources by headline volume and merge all remaining sources into an `Other` group.

This preserves the main source-level signals while reducing dimensionality.

In [16]:
TOP_N = 12

top_sources = (
    raw_scores["source"]
    .value_counts()
    .head(TOP_N)
    .index
    .tolist()
)

print(f"Top {TOP_N} sources:")
for source in top_sources:
    print("-", source)

raw_scores_top = raw_scores.copy()

raw_scores_top["source_grouped"] = np.where(
    raw_scores_top["source"].isin(top_sources),
    raw_scores_top["source"],
    "Other",
)

per_group_sum = (
    raw_scores_top
    .groupby(["date", "source_grouped"], observed=True)[SCORE_COLS]
    .sum()
    .reset_index()
)

per_group_count = (
    raw_scores_top
    .groupby(["date", "source_grouped"], observed=True)
    .size()
    .reset_index(name="count")
)

per_group_long = per_group_sum.merge(
    per_group_count,
    on=["date", "source_grouped"],
    how="left",
)

group_pivots = []

for col in [*SCORE_COLS, "count"]:
    pivot = per_group_long.pivot(
        index="date",
        columns="source_grouped",
        values=col,
    )

    pivot.columns = [
        f"{col}_{safe_col(source)}"
        for source in pivot.columns
    ]

    group_pivots.append(pivot)

top_source_wide = (
    pd.concat(group_pivots, axis=1)
    .sort_index()
    .fillna(0)
)

print("top_source_wide shape:", top_source_wide.shape)
top_source_wide.head()

Top 12 sources:
- N12 - חדשות
- סרוגים
- מעריב
- N12 - דף הבית
- וואלה! חדשות
- וואלה! חדשות - מבזקים
- כיפה - חדשות
- הארץ - מבזקים
- Ynet - חדשות
- חדשות 0404
- ישראל היום - חדשות
- הארץ - חדשות
top_source_wide shape: (3771, 104)


,relevance_politics_N12_-_דף_הבית,relevance_politics_N12_-_חדשות,relevance_politics_Other,relevance_politics_Ynet_-_חדשות,relevance_politics_הארץ_-_חדשות,relevance_politics_הארץ_-_מבזקים,relevance_politics_וואלה__חדשות,relevance_politics_וואלה__חדשות_-_מבזקים,relevance_politics_חדשות_0404,relevance_politics_ישראל_היום_-_חדשות,relevance_politics_כיפה_-_חדשות,relevance_politics_מעריב,relevance_politics_סרוגים,relevance_economy_N12_-_דף_הבית,relevance_economy_N12_-_חדשות,relevance_economy_Other,relevance_economy_Ynet_-_חדשות,relevance_economy_הארץ_-_חדשות,relevance_economy_הארץ_-_מבזקים,relevance_economy_וואלה__חדשות,relevance_economy_וואלה__חדשות_-_מבזקים,relevance_economy_חדשות_0404,relevance_economy_ישראל_היום_-_חדשות,relevance_economy_כיפה_-_חדשות,relevance_economy_מעריב,relevance_economy_סרוגים,relevance_security_N12_-_דף_הבית,relevance_security_N12_-_חדשות,relevance_security_Other,relevance_security_Ynet_-_חדשות,relevance_security_הארץ_-_חדשות,relevance_security_הארץ_-_מבזקים,relevance_security_וואלה__חדשות,relevance_security_וואלה__חדשות_-_מבזקים,relevance_security_חדשות_0404,relevance_security_ישראל_היום_-_חדשות,relevance_security_כיפה_-_חדשות,relevance_security_מעריב,relevance_security_סרוגים,relevance_health_N12_-_דף_הבית,...,relevance_science_סרוגים,relevance_technology_N12_-_דף_הבית,relevance_technology_N12_-_חדשות,relevance_technology_Other,relevance_technology_Ynet_-_חדשות,relevance_technology_הארץ_-_חדשות,relevance_technology_הארץ_-_מבזקים,relevance_technology_וואלה__חדשות,relevance_technology_וואלה__חדשות_-_מבזקים,relevance_technology_חדשות_0404,relevance_technology_ישראל_היום_-_חדשות,relevance_technology_כיפה_-_חדשות,relevance_technology_מעריב,relevance_technology_סרוגים,global_sentiment_N12_-_דף_הבית,global_sentiment_N12_-_חדשות,global_sentiment_Other,global_sentiment_Ynet_-_חדשות,global_sentiment_הארץ_-_חדשות,global_sentiment_הארץ_-_מבזקים,global_sentiment_וואלה__חדשות,global_sentiment_וואלה__חדשות_-_מבזקים,global_sentiment_חדשות_0404,global_sentiment_ישראל_היום_-_חדשות,global_sentiment_כיפה_-_חדשות,global_sentiment_מעריב,global_sentiment_סרוגים,count_N12_-_דף_הבית,count_N12_-_חדשות,count_Other,count_Ynet_-_חדשות,count_הארץ_-_חדשות,count_הארץ_-_מבזקים,count_וואלה__חדשות,count_וואלה__חדשות_-_מבזקים,count_חדשות_0404,count_ישראל_היום_-_חדשות,count_כיפה_-_חדשות,count_מעריב,count_סרוגים
date,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
2015-12-17,119.0,264.0,1252.0,0.0,143.0,202.0,0.0,0.0,0.0,0.0,58.0,189.0,196.0,67.0,131.0,364.0,0.0,53.0,35.0,0.0,0.0,0.0,0.0,0.0,78.0,24.0,80.0,169.0,941.0,0.0,83.0,147.0,0.0,0.0,0.0,0.0,37.0,94.0,76.0,20.0,...,0.0,2.0,18.0,86.0,0.0,21.0,14.0,0.0,0.0,0.0,0.0,0.0,6.0,0.0,-49.0,-116.0,-353.0,0.0,-51.0,-104.0,0.0,0.0,0.0,0.0,-3.0,-41.0,5.0,32.0,71.0,265.0,0.0,41.0,36.0,0.0,0.0,0.0,0.0,10.0,56.0,32.0
2015-12-18,57.0,164.0,645.0,0.0,111.0,251.0,0.0,0.0,0.0,0.0,25.0,175.0,52.0,4.0,3.0,49.0,0.0,0.0,42.0,0.0,0.0,0.0,0.0,0.0,79.0,0.0,35.0,93.0,330.0,0.0,43.0,148.0,0.0,0.0,0.0,0.0,16.0,51.0,13.0,0.0,...,0.0,4.0,3.0,17.0,0.0,10.0,9.0,0.0,0.0,0.0,0.0,0.0,14.0,0.0,-21.0,-33.0,-178.0,0.0,-52.0,-134.0,0.0,0.0,0.0,0.0,-10.0,13.0,10.0,20.0,38.0,132.0,0.0,23.0,41.0,0.0,0.0,0.0,0.0,3.0,44.0,8.0
2015-12-19,54.0,101.0,489.0,0.0,64.0,161.0,0.0,0.0,0.0,0.0,2.0,193.0,51.0,10.0,7.0,49.0,0.0,11.0,3.0,0.0,0.0,0.0,0.0,0.0,41.0,0.0,25.0,70.0,436.0,0.0,28.0,113.0,0.0,0.0,0.0,0.0,9.0,106.0,18.0,9.0,...,0.0,7.0,16.0,19.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,13.0,0.0,-22.0,-15.0,-272.0,0.0,-14.0,-97.0,0.0,0.0,0.0,0.0,-9.0,-81.0,-23.0,18.0,28.0,110.0,0.0,20.0,29.0,0.0,0.0,0.0,0.0,2.0,49.0,10.0
2015-12-20,150.0,311.0,1166.0,0.0,175.0,365.0,0.0,0.0,0.0,0.0,47.0,281.0,194.0,28.0,63.0,171.0,0.0,22.0,27.0,0.0,0.0,0.0,0.0,0.0,47.0,7.0,111.0,214.0,1096.0,0.0,91.0,276.0,0.0,0.0,0.0,0.0,35.0,147.0,101.0,3.0,...,0.0,0.0,13.0,50.0,0.0,15.0,7.0,0.0,0.0,0.0,0.0,0.0,18.0,0.0,-32.0,-118.0,-642.0,0.0,-68.0,-201.0,0.0,0.0,0.0,0.0,-16.0,-28.0,-58.0,30.0,71.0,265.0,0.0,36.0,68.0,0.0,0.0,0.0,0.0,7.0,62.0,3

In [17]:
# Attach top-source news features to the next TA-125 trading day
news_dates_top = top_source_wide.index.values

positions_top = np.searchsorted(trading_arr, news_dates_top, side="left")
mask_top = positions_top < len(trading_arr)

top_news_attached = top_source_wide.iloc[mask_top].copy()
top_news_attached["trading_day"] = trading_days[positions_top[mask_top]]

top_news_per_trading_day = (
    top_news_attached
    .groupby("trading_day")
    .sum()
    .sort_index()
)

print("Top-source news per trading day:", top_news_per_trading_day.shape)


# Rebuild merged dataset using top-source news features
merged_top = pd.DataFrame(index=trading_days)
merged_top.index.name = "date"

merged_top = (
    merged_top
    .join(ta125_clean, how="left")
    .join(vta35_clean, how="left")
    .join(market, how="left")
    .join(fx, how="left")
    .join(top_news_per_trading_day, how="left")
)

market_like_cols_top = [
    col for col in merged_top.columns
    if col.startswith(("Market_", "FX_", "VTA35_"))
]

top_news_cols = top_news_per_trading_day.columns.tolist()

merged_top[market_like_cols_top] = merged_top[market_like_cols_top].ffill()
merged_top[top_news_cols] = merged_top[top_news_cols].fillna(0)

merged_top = merged_top.sort_index().copy()

next_ta125_price_top = merged_top["TA125_Price"].shift(-1)

merged_top["Target"] = np.where(
    next_ta125_price_top.isna(),
    np.nan,
    (next_ta125_price_top > merged_top["TA125_Price"]).astype(int),
)

merged_top = merged_top.drop(columns=["TA125_Price"])

merged_top = add_vta35_availability(merged_top)

df_ml_top = (
    merged_top
    .replace([np.inf, -np.inf], np.nan)
    .dropna()
    .copy()
)

X_top = df_ml_top.drop(columns=["Target"])
y_top = df_ml_top["Target"].astype(int)

print("Top-source final dataset shape:", df_ml_top.shape)
print("Top-source feature count:", X_top.shape[1])
print("Date range:", df_ml_top.index.min().date(), "->", df_ml_top.index.max().date())
print()
print("Target balance:")
print(y_top.value_counts(normalize=True).rename({0: "Fall", 1: "Rise"}))

Top-source news per trading day: (2534, 104)
Top-source final dataset shape: (2516, 112)
Top-source feature count: 111
Date range: 2015-12-17 -> 2026-04-28

Target balance:
Target
Rise    0.538553
Fall    0.461447
Name: proportion, dtype: float64


## 16. Train POC Models on Top-Source Features

We now train the same three POC models on the reduced top-source feature set.



In [18]:
split_idx_top = int(len(df_ml_top) * 0.8)

X_train_top = X_top.iloc[:split_idx_top]
X_test_top = X_top.iloc[split_idx_top:]

y_train_top = y_top.iloc[:split_idx_top]
y_test_top = y_top.iloc[split_idx_top:]

print(f"Train rows: {len(X_train_top):,}")
print(f"Test rows:  {len(X_test_top):,}")
print()
print(f"Train date range: {X_train_top.index.min().date()} -> {X_train_top.index.max().date()}")
print(f"Test date range:  {X_test_top.index.min().date()} -> {X_test_top.index.max().date()}")
print()
print(f"Train rise rate: {y_train_top.mean():.2%}")
print(f"Test rise rate:  {y_test_top.mean():.2%}")

models_top = {
    "XGBoost": XGBClassifier(
        n_estimators=200,
        max_depth=4,
        learning_rate=0.05,
        subsample=0.8,
        colsample_bytree=0.8,
        eval_metric="logloss",
        random_state=SEED,
    ),
    "LGBM": LGBMClassifier(
        n_estimators=200,
        learning_rate=0.03,
        num_leaves=31,
        subsample=0.8,
        colsample_bytree=0.8,
        random_state=SEED,
        verbose=-1,
    ),
    "CatBoost": CatBoostClassifier(
        iterations=200,
        learning_rate=0.03,
        depth=6,
        l2_leaf_reg=3,
        verbose=0,
        random_seed=SEED,
    ),
}

predictions_top = {}

for name, model in models_top.items():
    print(f"\nTraining {name}...")

    model.fit(X_train_top, y_train_top)

    y_pred = model.predict(X_test_top)
    y_proba = model.predict_proba(X_test_top)[:, 1]

    predictions_top[name] = {
        "model": model,
        "y_pred": y_pred,
        "y_proba": y_proba,
    }

    print(f"{name} accuracy: {accuracy_score(y_test_top, y_pred):.2%}")
    print(classification_report(y_test_top, y_pred, target_names=["Fall", "Rise"]))
    print("-" * 80)

Train rows: 2,012
Test rows:  504

Train date range: 2015-12-17 -> 2024-03-25
Test date range:  2024-03-26 -> 2026-04-28

Train rise rate: 53.13%
Test rise rate:  56.75%

Training XGBoost...
XGBoost accuracy: 57.14%
              precision    recall  f1-score   support

        Fall       0.52      0.11      0.19       218
        Rise       0.58      0.92      0.71       286

    accuracy                           0.57       504
   macro avg       0.55      0.52      0.45       504
weighted avg       0.55      0.57      0.48       504

--------------------------------------------------------------------------------

Training LGBM...
LGBM accuracy: 57.94%
              precision    recall  f1-score   support

        Fall       0.57      0.11      0.18       218
        Rise       0.58      0.94      0.72       286

    accuracy                           0.58       504
   macro avg       0.58      0.52      0.45       504
weighted avg       0.58      0.58      0.48       504

---------

## 15. Statistical Evaluation for Top-Source Experiment

We now evaluate the top-source experiment using the same statistical tests as before.

This keeps the comparison fair:

- Same test period
- Same baseline logic
- Same metrics
- Same statistical tests

In [19]:
baseline_class_top = int(y_train_top.mean() > 0.5)
baseline_acc_top = float((y_test_top == baseline_class_top).mean())

print(f"Majority baseline class: {baseline_class_top}")
print(f"Majority baseline accuracy: {baseline_acc_top:.2%}")

rng = np.random.default_rng(SEED)

rows_top = []

y_test_top_arr = y_test_top.values
n_test_top = len(y_test_top_arr)

for name, pred_data in predictions_top.items():
    y_pred = pred_data["y_pred"]
    y_proba = pred_data["y_proba"]

    acc = float(accuracy_score(y_test_top_arr, y_pred))
    bal_acc = float(balanced_accuracy_score(y_test_top_arr, y_pred))
    n_correct = int((y_pred == y_test_top_arr).sum())

    p_binom = binomtest(
        n_correct,
        n_test_top,
        p=baseline_acc_top,
        alternative="greater",
    ).pvalue

    correct_mask = y_pred == y_test_top_arr

    boot_accs = np.empty(1000)
    for k in range(1000):
        idx = rng.integers(0, n_test_top, n_test_top)
        boot_accs[k] = correct_mask[idx].mean()

    ci_low, ci_high = np.percentile(boot_accs, [2.5, 97.5])

    perm_accs = np.empty(1000)
    for k in range(1000):
        shuffled = rng.permutation(y_test_top_arr)
        perm_accs[k] = (y_pred == shuffled).mean()

    p_perm = float((perm_accs >= acc).mean())

    fpr, tpr, _ = roc_curve(y_test_top_arr, y_proba)
    roc_auc = float(auc(fpr, tpr))

    rows_top.append({
        "Experiment": "Top sources + Other",
        "Aggregation": f"Top {TOP_N} sources + Other",
        "Feature Set": "Reduced per-source news + market basics",
        "Tuning": "POC default parameters",
        "Model": name,
        "Accuracy": acc,
        "Balanced Accuracy": bal_acc,
        "Baseline": baseline_acc_top,
        "Gap vs Baseline": acc - baseline_acc_top,
        "p_binom": p_binom,
        "p_perm": p_perm,
        "CI_low": ci_low,
        "CI_high": ci_high,
        "ROC_AUC": roc_auc,
    })

results_top_df = (
    pd.DataFrame(rows_top)
    .sort_values("Accuracy", ascending=False)
    .reset_index(drop=True)
)

results_top_df

Majority baseline class: 1
Majority baseline accuracy: 56.75%


,Experiment,Aggregation,Feature Set,Tuning,Model,Accuracy,Balanced Accuracy,Baseline,Gap vs Baseline,p_binom,p_perm,CI_low,CI_high,ROC_AUC
0,Top sources + Other,Top 12 sources + Other,Reduced per-source news + market basics,POC default parameters,LGBM,0.579365,0.523032,0.56746,0.011905,0.311039,0.052,0.533730,0.623016,0.541509
1,Top sources + Other,Top 12 sources + Other,Reduced per-source news + market basics,POC default parameters,XGBoost,0.571429,0.517130,0.56746,0.003968,0.447154,0.130,0.527778,0.615079,0.535895
2,Top sources + Other,Top 12 sources + Other,Reduced per-source news + market basics,POC default parameters,CatBoost,0.567460,0.506544,0.56746,0.000000,0.518730,0.315,0.525794,0.609127,0.496327


In [20]:
experiment_results = pd.concat(
    [experiment_results, results_top_df],
    ignore_index=True,
)

experiment_results.sort_values(
    ["Accuracy", "Balanced Accuracy"],
    ascending=False,
).reset_index(drop=True)

,Experiment,Aggregation,Feature Set,Tuning,Model,Accuracy,Balanced Accuracy,Baseline,Gap vs Baseline,p_binom,p_perm,CI_low,CI_high,ROC_AUC
0,Top sources + Other,Top 12 sources + Other,Reduced per-source news + market basics,POC default parameters,LGBM,0.579365,0.523032,0.56746,0.011905,0.311039,0.052,0.533730,0.623016,0.541509
1,Top sources + Other,Top 12 sources + Other,Reduced per-source news + market basics,POC default parameters,XGBoost,0.571429,0.517130,0.56746,0.003968,0.447154,0.130,0.527778,0.615079,0.535895
2,Baseline wide features,LSTM per-source wide sum,News per source + market basics,POC default parameters,CatBoost,0.571429,0.506768,0.56746,0.003968,0.447154,0.234,0.527778,0.613095,0.475974
3,Baseline wide features,LSTM per-source wide sum,News per source + market basics,POC default parameters,XGBoost,0.569444,0.507202,0.56746,0.001984,0.482880,0.255,0.525794,0.611161,0.519592
4,Baseline wide features,LSTM per-source wide sum,News per source + market basics,POC default parameters,LGBM,0.569444,0.506656,0.56746,0.001984,0.482880,0.278,0.525794,0.615079,0.545294
5,Top sources + Other,Top 12 sources + Other,Reduced per-source news + market basics,POC default parameters,CatBoost,0.567460,0.506544,0.56746,0.000000,0.518730,0.315,0.525794,0.609127,0.496327


## 17. Improvement Attempt 2: Market-Derived Features

The previous experiments used mostly raw market levels:

- TA-125 volume
- VTA-35 price
- S&P 500 level
- VIX level
- Brent Oil level
- USD/ILS level

For stock-index direction prediction, derived market features are often more useful than raw levels.

In this experiment, we add:

- Lagged TA-125 log returns
- 5-day mean return
- 5-day and 20-day volatility
- RSI(14)
- TA-125 volume z-score
- Day-of-week indicators

This follows the feature-engineering direction explored in `tuning.ipynb`.

In [21]:
def add_market_derived_features(df: pd.DataFrame, ta125_price: pd.Series) -> pd.DataFrame:
    """Add leak-safe TA-125 derived features.

    All features are based only on information available up to the current day.
    The target still predicts the next trading day.
    """
    out = df.copy()

    price = ta125_price.reindex(out.index).astype(float)

    log_price = np.log(price)
    logret = log_price.diff()

    # Lagged returns: previous return values available at the current day.
    for lag in range(1, 8):
        out[f"TA125_logret_lag{lag}"] = logret.shift(lag)

    out["TA125_logret_5d_mean"] = logret.shift(1).rolling(5).mean()
    out["TA125_logret_5d_std"] = logret.shift(1).rolling(5).std()
    out["TA125_logret_20d_std"] = logret.shift(1).rolling(20).std()

    # RSI(14), calculated from past returns only.
    delta = price.diff()
    gain = delta.clip(lower=0)
    loss = -delta.clip(upper=0)

    avg_gain = gain.shift(1).rolling(14).mean()
    avg_loss = loss.shift(1).rolling(14).mean()

    rs = avg_gain / avg_loss
    out["TA125_RSI14"] = 100 - (100 / (1 + rs))

    # Volume z-score using past 20 trading days.
    volume = out["TA125_Volume"].astype(float)
    volume_mean_20 = volume.shift(1).rolling(20).mean()
    volume_std_20 = volume.shift(1).rolling(20).std()
    out["TA125_volume_z20d"] = (volume - volume_mean_20) / volume_std_20

    # Day-of-week indicators.
    dow = pd.Series(out.index.dayofweek, index=out.index, name="dow")
    dow_dummies = pd.get_dummies(dow, prefix="DoW", drop_first=True).astype(int)
    out = out.join(dow_dummies)

    return out

## 18. Market-Derived Features

Add lagged returns, volatility, RSI, volume z-score, and day-of-week features.


In [22]:
# Rebuild the original wide-feature merged dataset.
merged_market = pd.DataFrame(index=trading_days)
merged_market.index.name = "date"

merged_market = (
    merged_market
    .join(ta125_clean, how="left")
    .join(vta35_clean, how="left")
    .join(market, how="left")
    .join(fx, how="left")
    .join(news_per_trading_day, how="left")
)

market_like_cols_market = [
    col for col in merged_market.columns
    if col.startswith(("Market_", "FX_", "VTA35_"))
]

news_cols_market = news_per_trading_day.columns.tolist()

merged_market[market_like_cols_market] = merged_market[market_like_cols_market].ffill()
merged_market[news_cols_market] = merged_market[news_cols_market].fillna(0)

merged_market = merged_market.sort_index().copy()

# Add derived market features before creating the final ML dataset.
merged_market = add_market_derived_features(
    merged_market,
    ta125_price=ta125_clean["TA125_Price"],
)

next_ta125_price_market = merged_market["TA125_Price"].shift(-1)

merged_market["Target"] = np.where(
    next_ta125_price_market.isna(),
    np.nan,
    (next_ta125_price_market > merged_market["TA125_Price"]).astype(int),
)

merged_market = merged_market.drop(columns=["TA125_Price"])

merged_market = add_vta35_availability(merged_market)

df_ml_market = (
    merged_market
    .replace([np.inf, -np.inf], np.nan)
    .dropna()
    .copy()
)

X_market = df_ml_market.drop(columns=["Target"])
y_market = df_ml_market["Target"].astype(int)

print("Market-derived dataset shape:", df_ml_market.shape)
print("Feature count:", X_market.shape[1])
print("Date range:", df_ml_market.index.min().date(), "->", df_ml_market.index.max().date())
print()
print("Target balance:")
print(y_market.value_counts(normalize=True).rename({0: "Fall", 1: "Rise"}))

new_market_features = [
    col for col in X_market.columns
    if col.startswith(("TA125_logret", "TA125_RSI", "TA125_volume_z", "DoW_"))
]

print()
print("New market-derived features:")
print(new_market_features)

Market-derived dataset shape: (2375, 345)
Feature count: 344
Date range: 2016-01-17 -> 2026-04-28

Target balance:
Target
Rise    0.539789
Fall    0.460211
Name: proportion, dtype: float64

New market-derived features:
['TA125_logret_lag1', 'TA125_logret_lag2', 'TA125_logret_lag3', 'TA125_logret_lag4', 'TA125_logret_lag5', 'TA125_logret_lag6', 'TA125_logret_lag7', 'TA125_logret_5d_mean', 'TA125_logret_5d_std', 'TA125_logret_20d_std', 'TA125_RSI14', 'TA125_volume_z20d', 'DoW_1', 'DoW_2', 'DoW_3', 'DoW_4', 'DoW_6']


In [23]:
merged_market = pd.DataFrame(index=trading_days)
merged_market.index.name = "date"

merged_market = (
    merged_market
    .join(ta125_clean, how="left")
    .join(vta35_clean, how="left")
    .join(market, how="left")
    .join(fx, how="left")
    .join(news_per_trading_day, how="left")
)

market_like_cols_market = [
    col for col in merged_market.columns
    if col.startswith(("Market_", "FX_", "VTA35_"))
]

news_cols_market = news_per_trading_day.columns.tolist()

merged_market[market_like_cols_market] = merged_market[market_like_cols_market].ffill()
merged_market[news_cols_market] = merged_market[news_cols_market].fillna(0)

merged_market = merged_market.sort_index().copy()

merged_market = add_market_derived_features(
    merged_market,
    ta125_price=ta125_clean["TA125_Price"],
)

next_ta125_price_market = merged_market["TA125_Price"].shift(-1)

merged_market["Target"] = np.where(
    next_ta125_price_market.isna(),
    np.nan,
    (next_ta125_price_market > merged_market["TA125_Price"]).astype(int),
)

merged_market = merged_market.drop(columns=["TA125_Price"])

merged_market = add_vta35_availability(merged_market)

df_ml_market = (
    merged_market
    .replace([np.inf, -np.inf], np.nan)
    .dropna()
    .copy()
)

X_market = df_ml_market.drop(columns=["Target"])
y_market = df_ml_market["Target"].astype(int)

print("Market-derived dataset shape:", df_ml_market.shape)
print("Feature count:", X_market.shape[1])
print("Date range:", df_ml_market.index.min().date(), "->", df_ml_market.index.max().date())
print()
print("Target balance:")
print(y_market.value_counts(normalize=True).rename({0: "Fall", 1: "Rise"}))

new_market_features = [
    col for col in X_market.columns
    if col.startswith(("TA125_logret", "TA125_RSI", "TA125_volume_z", "DoW_"))
]

print()
print("New market-derived features:")
print(new_market_features)

Market-derived dataset shape: (2375, 345)
Feature count: 344
Date range: 2016-01-17 -> 2026-04-28

Target balance:
Target
Rise    0.539789
Fall    0.460211
Name: proportion, dtype: float64

New market-derived features:
['TA125_logret_lag1', 'TA125_logret_lag2', 'TA125_logret_lag3', 'TA125_logret_lag4', 'TA125_logret_lag5', 'TA125_logret_lag6', 'TA125_logret_lag7', 'TA125_logret_5d_mean', 'TA125_logret_5d_std', 'TA125_logret_20d_std', 'TA125_RSI14', 'TA125_volume_z20d', 'DoW_1', 'DoW_2', 'DoW_3', 'DoW_4', 'DoW_6']


In [24]:
split_idx_market = int(len(df_ml_market) * 0.8)

X_train_market = X_market.iloc[:split_idx_market]
X_test_market = X_market.iloc[split_idx_market:]

y_train_market = y_market.iloc[:split_idx_market]
y_test_market = y_market.iloc[split_idx_market:]

print(f"Train rows: {len(X_train_market):,}")
print(f"Test rows:  {len(X_test_market):,}")
print()
print(f"Train date range: {X_train_market.index.min().date()} -> {X_train_market.index.max().date()}")
print(f"Test date range:  {X_test_market.index.min().date()} -> {X_test_market.index.max().date()}")
print()
print(f"Train rise rate: {y_train_market.mean():.2%}")
print(f"Test rise rate:  {y_test_market.mean():.2%}")

Train rows: 1,900
Test rows:  475

Train date range: 2016-01-17 -> 2024-01-09
Test date range:  2024-01-10 -> 2026-04-28

Train rise rate: 53.37%
Test rise rate:  56.42%


In [25]:
models_market = {
    "XGBoost": XGBClassifier(
        n_estimators=200,
        max_depth=4,
        learning_rate=0.05,
        subsample=0.8,
        colsample_bytree=0.8,
        eval_metric="logloss",
        random_state=SEED,
    ),
    "LGBM": LGBMClassifier(
        n_estimators=200,
        learning_rate=0.03,
        num_leaves=31,
        subsample=0.8,
        colsample_bytree=0.8,
        random_state=SEED,
        verbose=-1,
    ),
    "CatBoost": CatBoostClassifier(
        iterations=200,
        learning_rate=0.03,
        depth=6,
        l2_leaf_reg=3,
        verbose=0,
        random_seed=SEED,
    ),
}

predictions_market = {}

for name, model in models_market.items():
    print(f"\nTraining {name}...")

    model.fit(X_train_market, y_train_market)

    y_pred = model.predict(X_test_market)
    y_proba = model.predict_proba(X_test_market)[:, 1]

    predictions_market[name] = {
        "model": model,
        "y_pred": y_pred,
        "y_proba": y_proba,
    }

    print(f"{name} accuracy: {accuracy_score(y_test_market, y_pred):.2%}")
    print(classification_report(y_test_market, y_pred, target_names=["Fall", "Rise"]))
    print("-" * 80)


Training XGBoost...
XGBoost accuracy: 57.05%
              precision    recall  f1-score   support

        Fall       0.53      0.13      0.21       207
        Rise       0.58      0.91      0.71       268

    accuracy                           0.57       475
   macro avg       0.55      0.52      0.46       475
weighted avg       0.56      0.57      0.49       475

--------------------------------------------------------------------------------

Training LGBM...
LGBM accuracy: 57.05%
              precision    recall  f1-score   support

        Fall       0.52      0.18      0.27       207
        Rise       0.58      0.87      0.70       268

    accuracy                           0.57       475
   macro avg       0.55      0.53      0.48       475
weighted avg       0.55      0.57      0.51       475

--------------------------------------------------------------------------------

Training CatBoost...
CatBoost accuracy: 57.68%
              precision    recall  f1-score   supp

In [26]:
baseline_class_market = int(y_train_market.mean() > 0.5)
baseline_acc_market = float((y_test_market == baseline_class_market).mean())

print(f"Majority baseline class: {baseline_class_market}")
print(f"Majority baseline accuracy: {baseline_acc_market:.2%}")

rng = np.random.default_rng(SEED)
rows_market = []

y_true = y_test_market.values
n_test = len(y_true)

for name, pred_data in predictions_market.items():
    y_pred = pred_data["y_pred"]
    y_proba = pred_data["y_proba"]

    acc = float(accuracy_score(y_true, y_pred))
    bal_acc = float(balanced_accuracy_score(y_true, y_pred))
    n_correct = int((y_pred == y_true).sum())

    p_binom = binomtest(
        n_correct,
        n_test,
        p=baseline_acc_market,
        alternative="greater",
    ).pvalue

    correct_mask = y_pred == y_true

    boot_accs = np.empty(1000)
    for k in range(1000):
        idx = rng.integers(0, n_test, n_test)
        boot_accs[k] = correct_mask[idx].mean()

    ci_low, ci_high = np.percentile(boot_accs, [2.5, 97.5])

    perm_accs = np.empty(1000)
    for k in range(1000):
        shuffled = rng.permutation(y_true)
        perm_accs[k] = (y_pred == shuffled).mean()

    p_perm = float((perm_accs >= acc).mean())

    fpr, tpr, _ = roc_curve(y_true, y_proba)
    roc_auc = float(auc(fpr, tpr))

    rows_market.append({
        "Experiment": "Market-derived features",
        "Aggregation": "LSTM per-source wide sum",
        "Feature Set": "News + market-derived features",
        "Tuning": "POC default parameters",
        "Model": name,
        "Accuracy": acc,
        "Balanced Accuracy": bal_acc,
        "Baseline": baseline_acc_market,
        "Gap vs Baseline": acc - baseline_acc_market,
        "p_binom": p_binom,
        "p_perm": p_perm,
        "CI_low": ci_low,
        "CI_high": ci_high,
        "ROC_AUC": roc_auc,
    })

results_market_df = pd.DataFrame(rows_market)
results_market_df = results_market_df.sort_values("Accuracy", ascending=False).reset_index(drop=True)

results_market_df

Majority baseline class: 1
Majority baseline accuracy: 56.42%


,Experiment,Aggregation,Feature Set,Tuning,Model,Accuracy,Balanced Accuracy,Baseline,Gap vs Baseline,p_binom,p_perm,CI_low,CI_high,ROC_AUC
0,Market-derived features,LSTM per-source wide sum,News + market-derived features,POC default parameters,CatBoost,0.576842,0.521090,0.564211,0.012632,0.305940,0.050,0.534737,0.623158,0.507409
1,Market-derived features,LSTM per-source wide sum,News + market-derived features,POC default parameters,XGBoost,0.570526,0.520441,0.564211,0.006316,0.409273,0.116,0.526316,0.616842,0.558169
2,Market-derived features,LSTM per-source wide sum,News + market-derived features,POC default parameters,LGBM,0.570526,0.525939,0.564211,0.006316,0.409273,0.080,0.526316,0.614789,0.566101


In [27]:
summary_table = (
    experiment_results
    .sort_values(["Gap vs Baseline", "Accuracy"], ascending=False)
    .reset_index(drop=True)
)

best_idx = summary_table["Gap vs Baseline"].idxmax()

def highlight_best(row):
    if row.name == best_idx:
        return ["background-color: #d1fae5; font-weight: bold"] * len(row)
    return [""] * len(row)

summary_table.style.apply(highlight_best, axis=1)

,Experiment,Aggregation,Feature Set,Tuning,Model,Accuracy,Balanced Accuracy,Baseline,Gap vs Baseline,p_binom,p_perm,CI_low,CI_high,ROC_AUC
0,Top sources + Other,Top 12 sources + Other,Reduced per-source news + market basics,POC default parameters,LGBM,0.579365,0.523032,0.567460,0.011905,0.311039,0.052000,0.533730,0.623016,0.541509
1,Baseline wide features,LSTM per-source wide sum,News per source + market basics,POC default parameters,CatBoost,0.571429,0.506768,0.567460,0.003968,0.447154,0.234000,0.527778,0.613095,0.475974
2,Top sources + Other,Top 12 sources + Other,Reduced per-source news + market basics,POC default parameters,XGBoost,0.571429,0.517130,0.567460,0.003968,0.447154,0.130000,0.527778,0.615079,0.535895
3,Baseline wide features,LSTM per-source wide sum,News per source + market basics,POC default parameters,XGBoost,0.569444,0.507202,0.567460,0.001984,0.482880,0.255000,0.525794,0.611161,0.519592
4,Baseline wide features,LSTM per-source wide sum,News per source + market basics,POC default parameters,LGBM,0.569444,0.506656,0.567460,0.001984,0.482880,0.278000,0.525794,0.615079,0.545294
5,Top sources + Other,Top 12 sources + Other,Reduced per-source news + market basics,POC default parameters,CatBoost,0.567460,0.506544,0.567460,0.000000,0.518730,0.315000,0.525794,0.609127,0.496327


## 20. Walk-Forward Validation

Validate the best current model across several future test windows.

This checks whether the result is stable over time, instead of depending on one lucky train/test split.

In [28]:
from sklearn.base import clone

def walk_forward_eval(
    model,
    X_data: pd.DataFrame,
    y_data: pd.Series,
    n_folds: int = 5,
    test_window: int = 60,
):
    rows = []

    n = len(X_data)
    total_test = n_folds * test_window
    first_test_start = n - total_test

    for fold in range(n_folds):
        test_start = first_test_start + fold * test_window
        test_end = test_start + test_window

        X_train_fold = X_data.iloc[:test_start]
        y_train_fold = y_data.iloc[:test_start]

        X_test_fold = X_data.iloc[test_start:test_end]
        y_test_fold = y_data.iloc[test_start:test_end]

        fold_model = clone(model)
        fold_model.fit(X_train_fold, y_train_fold)

        y_pred_fold = fold_model.predict(X_test_fold)

        acc = accuracy_score(y_test_fold, y_pred_fold)
        bal_acc = balanced_accuracy_score(y_test_fold, y_pred_fold)

        majority_class = int(y_train_fold.mean() > 0.5)
        baseline = float((y_test_fold == majority_class).mean())

        rows.append({
            "Fold": fold + 1,
            "Train Start": X_train_fold.index.min().date(),
            "Train End": X_train_fold.index.max().date(),
            "Test Start": X_test_fold.index.min().date(),
            "Test End": X_test_fold.index.max().date(),
            "Accuracy": acc,
            "Balanced Accuracy": bal_acc,
            "Baseline": baseline,
            "Gap vs Baseline": acc - baseline,
            "Test Rise Rate": y_test_fold.mean(),
        })

    return pd.DataFrame(rows)

In [29]:
best_catboost_market = CatBoostClassifier(
    iterations=200,
    learning_rate=0.03,
    depth=6,
    l2_leaf_reg=3,
    verbose=0,
    random_seed=SEED,
)

walk_forward_market = walk_forward_eval(
    best_catboost_market,
    X_market,
    y_market,
    n_folds=5,
    test_window=60,
)

walk_forward_market

,Fold,Train Start,Train End,Test Start,Test End,Accuracy,Balanced Accuracy,Baseline,Gap vs Baseline,Test Rise Rate
0,1,2016-01-17,2025-01-01,2025-01-02,2025-03-26,0.633333,0.633333,0.500000,0.133333,0.500000
1,2,2016-01-17,2025-03-26,2025-03-27,2025-06-25,0.650000,0.598901,0.650000,0.000000,0.650000
2,3,2016-01-17,2025-06-25,2025-06-26,2025-09-18,0.566667,0.566667,0.500000,0.066667,0.500000
3,4,2016-01-17,2025-09-18,2025-09-21,2026-01-23,0.616667,0.578306,0.683333,-0.066667,0.683333
4,5,2016-01-17,2026-01-23,2026-01-26,2026-04-28,0.516667,0.491071,0.533333,-0.016667,0.533333


In [30]:
walk_forward_summary = pd.DataFrame([{
    "Mean Accuracy": walk_forward_market["Accuracy"].mean(),
    "Mean Balanced Accuracy": walk_forward_market["Balanced Accuracy"].mean(),
    "Mean Baseline": walk_forward_market["Baseline"].mean(),
    "Mean Gap vs Baseline": walk_forward_market["Gap vs Baseline"].mean(),
    "Positive Gap Folds": (walk_forward_market["Gap vs Baseline"] > 0).sum(),
    "Total Folds": len(walk_forward_market),
}])

walk_forward_summary

,Mean Accuracy,Mean Balanced Accuracy,Mean Baseline,Mean Gap vs Baseline,Positive Gap Folds,Total Folds
0,0.596667,0.573656,0.573333,0.023333,2,5


The walk-forward results show that the model is not consistently better than the majority-class baseline across time. It outperforms the baseline in folds 2 and 3, but underperforms in folds 1, 4, and 5. This suggests that the signal is unstable and may depend strongly on the specific test period.

## 21. Ablation Study

In [45]:
def evaluate_catboost_feature_set(
    X_data: pd.DataFrame,
    y_data: pd.Series,
    feature_cols: list[str],
    experiment_name: str,
    random_seed: int = SEED,
):
    X_subset = X_data[feature_cols].copy()

    split_idx = int(len(X_subset) * 0.8)

    X_train_ab = X_subset.iloc[:split_idx]
    X_test_ab = X_subset.iloc[split_idx:]

    y_train_ab = y_data.iloc[:split_idx]
    y_test_ab = y_data.iloc[split_idx:]

    model = CatBoostClassifier(
        iterations=200,
        learning_rate=0.03,
        depth=6,
        l2_leaf_reg=3,
        verbose=0,
        random_seed=random_seed,
    )

    model.fit(X_train_ab, y_train_ab)

    y_pred = model.predict(X_test_ab)
    y_proba = model.predict_proba(X_test_ab)[:, 1]

    baseline_class = int(y_train_ab.mean() > 0.5)
    baseline_acc = float((y_test_ab == baseline_class).mean())

    acc = float(accuracy_score(y_test_ab, y_pred))
    bal_acc = float(balanced_accuracy_score(y_test_ab, y_pred))
    n_correct = int((y_pred == y_test_ab.values).sum())
    n_test = len(y_test_ab)

    p_binom = binomtest(
        n_correct,
        n_test,
        p=baseline_acc,
        alternative="greater",
    ).pvalue

    fpr, tpr, _ = roc_curve(y_test_ab, y_proba)
    roc_auc = float(auc(fpr, tpr))

    return {
        "Experiment": experiment_name,
        "N Features": len(feature_cols),
        "Accuracy": acc,
        "Balanced Accuracy": bal_acc,
        "Baseline": baseline_acc,
        "Gap vs Baseline": acc - baseline_acc,
        "p_binom": p_binom,
        "ROC_AUC": roc_auc,
    }

In [46]:
market_derived_cols = [
    col for col in X_market.columns
    if col.startswith(("TA125_logret", "TA125_RSI", "TA125_volume_z", "DoW_"))
]

basic_market_cols = [
    "TA125_Volume",
    "VTA35_Price",
    "Market_Brent_Oil",
    "Market_SP500",
    "Market_VIX",
    "FX_USD_ILS",
]

news_cols_wide = [
    col for col in X_market.columns
    if col not in basic_market_cols + market_derived_cols
]

print("News wide features:", len(news_cols_wide))
print("Basic market features:", len(basic_market_cols))
print("Market-derived features:", len(market_derived_cols))

News wide features: 321
Basic market features: 6
Market-derived features: 17


In [47]:
ablation_rows = []

ablation_rows.append(
    evaluate_catboost_feature_set(
        X_market,
        y_market,
        feature_cols=news_cols_wide,
        experiment_name="News wide only",
    )
)

ablation_rows.append(
    evaluate_catboost_feature_set(
        X_market,
        y_market,
        feature_cols=basic_market_cols,
        experiment_name="Basic market only",
    )
)

ablation_rows.append(
    evaluate_catboost_feature_set(
        X_market,
        y_market,
        feature_cols=market_derived_cols,
        experiment_name="Market-derived only",
    )
)

ablation_rows.append(
    evaluate_catboost_feature_set(
        X_market,
        y_market,
        feature_cols=list(X_market.columns),
        experiment_name="News + all market features",
    )
)

ablation_df = (
    pd.DataFrame(ablation_rows)
    .sort_values("Gap vs Baseline", ascending=False)
    .reset_index(drop=True)
)

ablation_df

,Experiment,N Features,Accuracy,Balanced Accuracy,Baseline,Gap vs Baseline,p_binom,ROC_AUC
0,Basic market only,6,0.581053,0.538566,0.564211,0.016842,0.244185,0.541459
1,News + all market features,344,0.576842,0.521090,0.564211,0.012632,0.305940,0.507409
2,Market-derived only,17,0.560000,0.545200,0.564211,-0.004211,0.592185,0.529562
3,News wide only,321,0.555789,0.502433,0.564211,-0.008421,0.662017,0.484146


## 22. CatBoost Grid Search

In [48]:
n = len(df_ml_market)

n_train = int(n * 0.70)
n_val = int(n * 0.15)

X_train_grid = X_market.iloc[:n_train]
y_train_grid = y_market.iloc[:n_train]

X_val_grid = X_market.iloc[n_train:n_train + n_val]
y_val_grid = y_market.iloc[n_train:n_train + n_val]

X_test_grid = X_market.iloc[n_train + n_val:]
y_test_grid = y_market.iloc[n_train + n_val:]

print(f"Train: {len(X_train_grid):,} | {X_train_grid.index.min().date()} -> {X_train_grid.index.max().date()}")
print(f"Val:   {len(X_val_grid):,} | {X_val_grid.index.min().date()} -> {X_val_grid.index.max().date()}")
print(f"Test:  {len(X_test_grid):,} | {X_test_grid.index.min().date()} -> {X_test_grid.index.max().date()}")

print()
print(f"Train rise rate: {y_train_grid.mean():.2%}")
print(f"Val rise rate:   {y_val_grid.mean():.2%}")
print(f"Test rise rate:  {y_test_grid.mean():.2%}")

Train: 1,662 | 2016-01-17 -> 2022-12-26
Val:   356 | 2022-12-27 -> 2024-10-08
Test:  357 | 2024-10-09 -> 2026-04-28

Train rise rate: 53.31%
Val rise rate:   53.65%
Test rise rate:  57.42%


In [49]:
from itertools import product

cat_grid = {
    "iterations": [100, 200, 400],
    "depth": [4, 6, 8],
    "learning_rate": [0.01, 0.03, 0.05],
    "l2_leaf_reg": [1, 3, 10],
}

grid_rows = []

for iterations, depth, learning_rate, l2_leaf_reg in product(
    cat_grid["iterations"],
    cat_grid["depth"],
    cat_grid["learning_rate"],
    cat_grid["l2_leaf_reg"],
):
    params = {
        "iterations": iterations,
        "depth": depth,
        "learning_rate": learning_rate,
        "l2_leaf_reg": l2_leaf_reg,
        "verbose": 0,
        "random_seed": SEED,
    }

    model = CatBoostClassifier(**params)
    model.fit(X_train_grid, y_train_grid)

    val_pred = model.predict(X_val_grid)
    val_proba = model.predict_proba(X_val_grid)[:, 1]

    val_acc = accuracy_score(y_val_grid, val_pred)
    val_bal_acc = balanced_accuracy_score(y_val_grid, val_pred)

    fpr, tpr, _ = roc_curve(y_val_grid, val_proba)
    val_auc = auc(fpr, tpr)

    grid_rows.append({
        **params,
        "val_accuracy": val_acc,
        "val_balanced_accuracy": val_bal_acc,
        "val_auc": val_auc,
    })

cat_grid_results = (
    pd.DataFrame(grid_rows)
    .sort_values(["val_balanced_accuracy", "val_accuracy"], ascending=False)
    .reset_index(drop=True)
)

cat_grid_results.head(10)

,iterations,depth,learning_rate,l2_leaf_reg,verbose,random_seed,val_accuracy,val_balanced_accuracy,val_auc
0,100,6,0.03,3,0,42,0.533708,0.537807,0.513438
1,200,4,0.01,1,0,42,0.536517,0.537125,0.523084
2,400,8,0.01,10,0,42,0.528090,0.537109,0.510075
3,200,4,0.05,3,0,42,0.525281,0.535729,0.528193
4,400,6,0.05,1,0,42,0.533708,0.534507,0.526924
5,200,4,0.03,10,0,42,0.528090,0.533809,0.513121
6,400,6,0.01,1,0,42,0.522472,0.531461,0.502967
7,400,4,0.01,1,0,42,0.525281,0.530779,0.523338
8,200,6,0.05,3,0,42,0.514045,0.527320,0.490941
9,200,6,0.05,10,0,42,0.519663,0.527193,0.491385


In [50]:
best_cat_params = cat_grid_results.iloc[0][
    ["iterations", "depth", "learning_rate", "l2_leaf_reg"]
].to_dict()

best_cat_params = {
    "iterations": int(best_cat_params["iterations"]),
    "depth": int(best_cat_params["depth"]),
    "learning_rate": float(best_cat_params["learning_rate"]),
    "l2_leaf_reg": float(best_cat_params["l2_leaf_reg"]),
    "verbose": 0,
    "random_seed": SEED,
}

best_cat_params

{'iterations': 100,
 'depth': 6,
 'learning_rate': 0.03,
 'l2_leaf_reg': 3.0,
 'verbose': 0,
 'random_seed': 42}

In [51]:
X_trainval_grid = pd.concat([X_train_grid, X_val_grid])
y_trainval_grid = pd.concat([y_train_grid, y_val_grid])

cat_tuned = CatBoostClassifier(**best_cat_params)
cat_tuned.fit(X_trainval_grid, y_trainval_grid)

y_pred_tuned = cat_tuned.predict(X_test_grid)
y_proba_tuned = cat_tuned.predict_proba(X_test_grid)[:, 1]

test_acc_tuned = accuracy_score(y_test_grid, y_pred_tuned)
test_bal_acc_tuned = balanced_accuracy_score(y_test_grid, y_pred_tuned)

baseline_class_grid = int(y_trainval_grid.mean() > 0.5)
baseline_acc_grid = float((y_test_grid == baseline_class_grid).mean())

print(f"Tuned CatBoost test accuracy:          {test_acc_tuned:.2%}")
print(f"Tuned CatBoost test balanced accuracy: {test_bal_acc_tuned:.2%}")
print(f"Majority baseline accuracy:            {baseline_acc_grid:.2%}")
print(f"Gap vs baseline:                       {test_acc_tuned - baseline_acc_grid:.2%}")
print()
print(classification_report(y_test_grid, y_pred_tuned, target_names=["Fall", "Rise"]))

Tuned CatBoost test accuracy:          56.30%
Tuned CatBoost test balanced accuracy: 53.19%
Majority baseline accuracy:            57.42%
Gap vs baseline:                       -1.12%

              precision    recall  f1-score   support

        Fall       0.48      0.32      0.39       152
        Rise       0.60      0.74      0.66       205

    accuracy                           0.56       357
   macro avg       0.54      0.53      0.52       357
weighted avg       0.55      0.56      0.54       357



In [52]:
y_true_grid = y_test_grid.values
n_test_grid = len(y_true_grid)

acc = float(accuracy_score(y_true_grid, y_pred_tuned))
bal_acc = float(balanced_accuracy_score(y_true_grid, y_pred_tuned))
n_correct = int((y_pred_tuned == y_true_grid).sum())

p_binom = binomtest(
    n_correct,
    n_test_grid,
    p=baseline_acc_grid,
    alternative="greater",
).pvalue

rng = np.random.default_rng(SEED)

correct_mask = y_pred_tuned == y_true_grid
boot_accs = np.empty(1000)

for k in range(1000):
    idx = rng.integers(0, n_test_grid, n_test_grid)
    boot_accs[k] = correct_mask[idx].mean()

ci_low, ci_high = np.percentile(boot_accs, [2.5, 97.5])

perm_accs = np.empty(1000)

for k in range(1000):
    shuffled = rng.permutation(y_true_grid)
    perm_accs[k] = (y_pred_tuned == shuffled).mean()

p_perm = float((perm_accs >= acc).mean())

fpr, tpr, _ = roc_curve(y_true_grid, y_proba_tuned)
roc_auc = float(auc(fpr, tpr))

results_tuned_df = pd.DataFrame([{
    "Experiment": "CatBoost grid search",
    "Aggregation": "LSTM per-source wide sum",
    "Feature Set": "News + market-derived features",
    "Tuning": "Grid search on validation",
    "Model": "CatBoost",
    "Accuracy": acc,
    "Balanced Accuracy": bal_acc,
    "Baseline": baseline_acc_grid,
    "Gap vs Baseline": acc - baseline_acc_grid,
    "p_binom": p_binom,
    "p_perm": p_perm,
    "CI_low": ci_low,
    "CI_high": ci_high,
    "ROC_AUC": roc_auc,
}])

results_tuned_df

,Experiment,Aggregation,Feature Set,Tuning,Model,Accuracy,Balanced Accuracy,Baseline,Gap vs Baseline,p_binom,p_perm,CI_low,CI_high,ROC_AUC
0,CatBoost grid search,LSTM per-source wide sum,News + market-derived features,Grid search on validation,CatBoost,0.563025,0.531916,0.57423,-0.011204,0.685671,0.122,0.509734,0.616246,0.53344


### Grid Search Interpretation

CatBoost grid search improved validation performance, but did not improve the final test result.

The tuned model underperformed the majority-class baseline on the test split.

This suggests that the validation-selected hyperparameters do not generalize reliably, and that the current predictive signal is sensitive to the chosen time period.

## 23. Feature Importance

In [53]:
best_model = predictions_market["CatBoost"]["model"]

feature_importance = pd.DataFrame({
    "Feature": X_market.columns,
    "Importance": best_model.get_feature_importance(),
})

feature_importance = (
    feature_importance
    .sort_values("Importance", ascending=False)
    .reset_index(drop=True)
)

feature_importance.head(30)

,Feature,Importance
0,TA125_logret_5d_std,2.679823
1,TA125_logret_lag5,2.428868
2,TA125_RSI14,2.037855
3,TA125_logret_lag1,1.991493
4,TA125_logret_5d_mean,1.663738
5,relevance_health_N12_-_חדשות,1.418964
6,relevance_economy_הארץ_-_מבזקים,1.377715
7,TA125_Volume,1.373676
8,TA125_logret_lag6,1.275369
9,relevance_health_מעריב,1.169606


In [54]:
def feature_group(feature: str) -> str:
    if feature.startswith("TA125_logret") or feature.startswith("TA125_RSI") or feature.startswith("TA125_volume_z") or feature.startswith("DoW_"):
        return "Market-derived"
    if feature in basic_market_cols:
        return "Basic market"
    return "News"

feature_importance["Group"] = feature_importance["Feature"].apply(feature_group)

group_importance = (
    feature_importance
    .groupby("Group", as_index=False)["Importance"]
    .sum()
    .sort_values("Importance", ascending=False)
)

group_importance["Share"] = group_importance["Importance"] / group_importance["Importance"].sum()

group_importance

,Group,Importance,Share
2,News,79.211020,0.792110
1,Market-derived,17.593677,0.175937
0,Basic market,3.195304,0.031953


### Feature Importance Interpretation

The best holdout model assigns most total importance to news-related features, while several of the strongest individual features are market-derived.

This supports the ablation finding: the model performs best when news signals are combined with market context.

However, because walk-forward validation was not stable, feature importance should be interpreted as exploratory rather than definitive.

## 24. Feature Selection by Importance

top_k_rows = []

for k in [20, 50, 100, len(X_market.columns)]:
    if k == len(X_market.columns):
        selected_features = list(X_market.columns)
        experiment_name = "All features"
    else:
        selected_features = feature_importance.head(k)["Feature"].tolist()
        experiment_name = f"Top {k} features"

    result = evaluate_catboost_feature_set(
        X_market,
        y_market,
        feature_cols=selected_features,
        experiment_name=experiment_name,
    )

    top_k_rows.append(result)

top_k_df = (
    pd.DataFrame(top_k_rows)
    .sort_values("Gap vs Baseline", ascending=False)
    .reset_index(drop=True)
)

top_k_df

## 25. Multi-Seed Robustness

In [55]:
SEEDS = [42, 123, 456, 789, 2024]

seed_rows = []

for seed in SEEDS:
    model = CatBoostClassifier(
        iterations=200,
        learning_rate=0.03,
        depth=6,
        l2_leaf_reg=3,
        verbose=0,
        random_seed=seed,
    )

    model.fit(X_train_market, y_train_market)

    y_pred_seed = model.predict(X_test_market)
    y_proba_seed = model.predict_proba(X_test_market)[:, 1]

    acc = float(accuracy_score(y_test_market, y_pred_seed))
    bal_acc = float(balanced_accuracy_score(y_test_market, y_pred_seed))

    fpr, tpr, _ = roc_curve(y_test_market, y_proba_seed)
    roc_auc = float(auc(fpr, tpr))

    seed_rows.append({
        "Seed": seed,
        "Accuracy": acc,
        "Balanced Accuracy": bal_acc,
        "Baseline": baseline_acc_market,
        "Gap vs Baseline": acc - baseline_acc_market,
        "ROC_AUC": roc_auc,
    })

seed_robustness_df = pd.DataFrame(seed_rows)

seed_robustness_summary = pd.DataFrame([{
    "Mean Accuracy": seed_robustness_df["Accuracy"].mean(),
    "Std Accuracy": seed_robustness_df["Accuracy"].std(),
    "Min Accuracy": seed_robustness_df["Accuracy"].min(),
    "Max Accuracy": seed_robustness_df["Accuracy"].max(),
    "Mean Gap vs Baseline": seed_robustness_df["Gap vs Baseline"].mean(),
    "Positive Gap Seeds": (seed_robustness_df["Gap vs Baseline"] > 0).sum(),
    "Total Seeds": len(seed_robustness_df),
}])

display(seed_robustness_df)
display(seed_robustness_summary)

,Seed,Accuracy,Balanced Accuracy,Baseline,Gap vs Baseline,ROC_AUC
0,42,0.576842,0.521090,0.564211,0.012632,0.507409
1,123,0.581053,0.520973,0.564211,0.016842,0.576808
2,456,0.566316,0.505714,0.564211,0.002105,0.546759
3,789,0.572632,0.515709,0.564211,0.008421,0.554600
4,2024,0.560000,0.502316,0.564211,-0.004211,0.559900


,Mean Accuracy,Std Accuracy,Min Accuracy,Max Accuracy,Mean Gap vs Baseline,Positive Gap Seeds,Total Seeds
0,0.571368,0.008368,0.56,0.581053,0.007158,4,5


The seed robustness check shows that the model usually performs slightly above the majority baseline, with positive gaps in 4 out of 5 seeds. However, the average improvement is small, only 0.72 percentage points, and performance varies across seeds. This suggests that the model has a weak and somewhat unstable signal rather than a robust predictive advantage.

## Final Experiment Summary


In [64]:
holdout_cols_final = [
    "Experiment",
    "Model",
    "Accuracy",
    "Baseline",
    "Gap vs Baseline",
    "Balanced Accuracy",
    "ROC_AUC",
    "p_binom",
    "p_perm",
]

main_holdout_results_small = main_holdout_results[
    holdout_cols_final
].copy()

main_holdout_results_small

,Experiment,Model,Accuracy,Baseline,Gap vs Baseline,Balanced Accuracy,ROC_AUC,p_binom,p_perm
0,Top sources + Other,LGBM,0.579365,0.56746,0.011905,0.523032,0.541509,0.311039,0.052
1,Baseline wide features,CatBoost,0.571429,0.56746,0.003968,0.506768,0.475974,0.447154,0.234
2,Top sources + Other,XGBoost,0.571429,0.56746,0.003968,0.517130,0.535895,0.447154,0.130
3,Baseline wide features,XGBoost,0.569444,0.56746,0.001984,0.507202,0.519592,0.482880,0.255
4,Baseline wide features,LGBM,0.569444,0.56746,0.001984,0.506656,0.545294,0.482880,0.278
5,Top sources + Other,CatBoost,0.567460,0.56746,0.000000,0.506544,0.496327,0.518730,0.315


In [62]:
robustness_cols = [
    "Experiment",
    "Model",
    "Validation Type",
    "Mean Accuracy",
    "Mean Balanced Accuracy",
    "Mean Baseline",
    "Mean Gap vs Baseline",
    "Positive Gap Folds",
    "Total Folds",
    "Positive Gap Seeds",
    "Total Seeds",
]

robustness_results_small = robustness_results[
    [col for col in robustness_cols if col in robustness_results.columns]
].copy()

robustness_results_small

,Experiment,Model,Validation Type,Mean Accuracy,Mean Baseline
0,Walk-forward validation,CatBoost,walk-forward,0.596667,0.573333
1,Multi-seed robustness,CatBoost,multi-seed,0.571368,0.564211


In [66]:
holdout_cols_bootstrap = [
    "Experiment",
    "Model",
    "Accuracy",
    "Baseline",
    "Gap vs Baseline",
    "Balanced Accuracy",
    "ROC_AUC",
    "p_perm",
    "CI_low",
    "CI_high",
]

main_holdout_results_bootstrap = main_holdout_results[
    holdout_cols_bootstrap
].copy()

main_holdout_results_bootstrap["Bootstrap 95% CI"] = (
    main_holdout_results_bootstrap["CI_low"].round(3).astype(str)
    + " - "
    + main_holdout_results_bootstrap["CI_high"].round(3).astype(str)
)

main_holdout_results_bootstrap = main_holdout_results_bootstrap.drop(
    columns=["CI_low", "CI_high"]
)

main_holdout_results_bootstrap

,Experiment,Model,Accuracy,Baseline,Gap vs Baseline,Balanced Accuracy,ROC_AUC,p_perm,Bootstrap 95% CI
0,Top sources + Other,LGBM,0.579365,0.56746,0.011905,0.523032,0.541509,0.052,0.534 - 0.623
1,Baseline wide features,CatBoost,0.571429,0.56746,0.003968,0.506768,0.475974,0.234,0.528 - 0.613
2,Top sources + Other,XGBoost,0.571429,0.56746,0.003968,0.517130,0.535895,0.130,0.528 - 0.615
3,Baseline wide features,XGBoost,0.569444,0.56746,0.001984,0.507202,0.519592,0.255,0.526 - 0.611
4,Baseline wide features,LGBM,0.569444,0.56746,0.001984,0.506656,0.545294,0.278,0.526 - 0.615
5,Top sources + Other,CatBoost,0.567460,0.56746,0.000000,0.506544,0.496327,0.315,0.526 - 0.609


In [67]:
main_holdout_results_bootstrap = main_holdout_results_bootstrap[
    [
        "Experiment",
        "Model",
        "Accuracy",
        "Bootstrap 95% CI",
        "Baseline",
        "Gap vs Baseline",
        "Balanced Accuracy",
        "ROC_AUC",
        "p_perm",
    ]
]

main_holdout_results_bootstrap

,Experiment,Model,Accuracy,Bootstrap 95% CI,Baseline,Gap vs Baseline,Balanced Accuracy,ROC_AUC,p_perm
0,Top sources + Other,LGBM,0.579365,0.534 - 0.623,0.56746,0.011905,0.523032,0.541509,0.052
1,Baseline wide features,CatBoost,0.571429,0.528 - 0.613,0.56746,0.003968,0.506768,0.475974,0.234
2,Top sources + Other,XGBoost,0.571429,0.528 - 0.615,0.56746,0.003968,0.517130,0.535895,0.130
3,Baseline wide features,XGBoost,0.569444,0.526 - 0.611,0.56746,0.001984,0.507202,0.519592,0.255
4,Baseline wide features,LGBM,0.569444,0.526 - 0.615,0.56746,0.001984,0.506656,0.545294,0.278
5,Top sources + Other,CatBoost,0.567460,0.526 - 0.609,0.56746,0.000000,0.506544,0.496327,0.315
